{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "697cd71c-feab-4751-9b9f-c484d54d7568",
   "metadata": {},
   "source": [
    "CD4 Naive ATAC "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "562824eb-904d-48c0-aa01-8d1bec5f7760",
   "metadata": {},
   "outputs": [],
   "source": [
    "quiet_library <- function(...) { suppressPackageStartupMessages(library(...)) }\n",
    "quiet_library(ArchR)\n",
    "quiet_library(hise)\n",
    "quiet_library(Seurat)\n",
    "quiet_library(ggplot2)\n",
    "quiet_library(Matrix)\n",
    "quiet_library(H5weaver)\n",
    "quiet_library(dplyr)\n",
    "#quiet_library(viridis)\n",
    "#quiet_library(harmony)\n",
    "#quiet_library(Nebulosa)\n",
    "quiet_library(stringr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6db9b68-999e-4d1c-a817-6a744901bf2a",
   "metadata": {},
   "outputs": [],
   "source": [
    "addArchRThreads(32)\n",
    "addArchRGenome(\"hg38\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17f58c6c-43ad-494b-b52e-0af1ff029030",
   "metadata": {},
   "source": [
    "# Load Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c742969-04b4-43c8-a6ef-ffbe8ef70593",
   "metadata": {},
   "outputs": [],
   "source": [
    "proj <- loadArchRProject(path = '../../PedSen_ATAC/')\n",
    "proj"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59950829-bdb2-48d5-a9d8-71a81535adf9",
   "metadata": {},
   "source": [
    "## Subset ArchR Project"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a460e552-bd48-4a8a-a7cc-446d5eb78c52",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd4_na <- readRDS(file = 'cd4_na_labeled.rds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4de58ee6-2948-462d-be53-fdc877bf7f01",
   "metadata": {},
   "outputs": [],
   "source": [
    "bcs <- colnames(cd4_na[['RNA']])\n",
    "head(bcs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c999d9f4-503d-44d4-9636-e6ee49dec820",
   "metadata": {},
   "outputs": [],
   "source": [
    "idx_new <- which(proj$cellNames_clean %in% bcs)\n",
    "head(idx_new)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e531828f-fd7a-4cd0-820d-4d9e2504d254",
   "metadata": {},
   "outputs": [],
   "source": [
    "cells_subset <- proj$cellNames[idx_new]\n",
    "proj_subset <- proj[cells_subset,]\n",
    "proj_subset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f92caca8-6dd1-47d7-a5d1-4507e9b8951b",
   "metadata": {},
   "source": [
    "## Add Pure Celltypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50e145de-654e-4da5-8b19-aeeb2c31e893",
   "metadata": {},
   "outputs": [],
   "source": [
    "pure_clusters <- as.data.frame(cd4_na$pure_celltype)\n",
    "rownames(pure_clusters) <- paste(cd4_na$batch_id, \"-P1_\",cd4_na$pbmc_sample_id,\"#\",rownames(pure_clusters), sep = \"\")\n",
    "colnames(pure_clusters) <- 'pure_clusters'\n",
    "pure_clusters[] <- lapply(pure_clusters, as.character)\n",
    "head(pure_clusters)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2030cf43-fc01-49da-8117-388240fda38f",
   "metadata": {},
   "outputs": [],
   "source": [
    "proj_subset <- addCellColData(proj_subset, data = pure_clusters$pure_clusters, name = 'pure_clusters', cells = rownames(pure_clusters), force = TRUE)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e64582f-1c02-4f93-972a-8dc6af2f9094",
   "metadata": {},
   "source": [
    "# ATAC Clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6397df55-68a5-409b-b186-cf491baa8670",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "proj_subset <- addIterativeLSI(proj_subset, name = 'IterativeLSI', force = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "098429d2-674e-443e-b8ad-297e15024f01",
   "metadata": {},
   "outputs": [],
   "source": [
    "proj_subset <- addHarmony(proj_subset, reducedDims = 'IterativeLSI', groupBy = 'batch_id', force = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7664c36f-c086-4b8e-bc83-247e567ec4ff",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "#proj_subset <- addClusters(proj_subset, reducedDims = 'IterativeLSI', name = 'Clusters', force = TRUE, resolution = 0.2)\n",
    "proj_subset <- addClusters(proj_subset, reducedDims = 'Harmony', name = 'Clusters', force = TRUE, resolution = 0.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10b95edf-c8dc-41e2-9f04-a66273465ad4",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "#proj_subset <- addUMAP(proj_subset, reducedDims = 'IterativeLSI', name = 'UMAP', force = TRUE)\n",
    "proj_subset <- addUMAP(proj_subset, reducedDims = 'Harmony', name = 'UMAP', force = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bceeed6d-3d9d-4b66-b881-e4b14c92f91e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "options(repr.plocd4_naidth = 8, repr.plot.height = 8)\n",
    "plotEmbedding(ArchRProj = proj_subset, colorBy = \"cellColData\", name = \"Clusters\", embedding = \"UMAP\")\n",
    "plotEmbedding(ArchRProj = proj_subset, colorBy = \"cellColData\", name = \"pure_clusters\", embedding = \"UMAP\")\n",
    "plotEmbedding(ArchRProj = proj_subset, colorBy = \"cellColData\", name = \"pediatric_senior\", embedding = \"UMAP\")\n",
    "plotEmbedding(ArchRProj = proj_subset, colorBy = \"cellColData\", name = \"predicted.t_celltype.l2\", embedding = \"UMAP\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e5c4a0b-697b-487e-9470-3ebf349a72fe",
   "metadata": {},
   "source": [
    "# GS Plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6720ea3a-2e9e-4c24-bf96-bc3685cdb6f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "proj_subset <- addImputeWeights(proj_subset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d003000-a378-42b6-a176-ab9c2f0be6ee",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "markerGenes  <- c('SOX4','TOX','IKZF2','CPQ','STAT4','IFNG')\n",
    "\n",
    "p <- plotEmbedding(\n",
    "    ArchRProj = proj_subset, \n",
    "    colorBy = \"GeneScoreMatrix\", \n",
    "    name = markerGenes, \n",
    "    embedding = \"UMAP\",\n",
    "    imputeWeights = NULL\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd2c7343-1dab-473a-ac00-ddc3cb0365ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Rearrange for grid plotting\n",
    "options(repr.plot.width = 20, repr.plot.height = 16)\n",
    "p2 <- lapply(p, function(x){\n",
    "    x + guides(color = FALSE, fill = FALSE) + \n",
    "    theme_ArchR(baseSize = 6.5) +\n",
    "    theme(plot.margin = unit(c(0, 0, 0, 0), \"cm\")) +\n",
    "    theme(\n",
    "        axis.text.x=element_blank(), \n",
    "        axis.ticks.x=element_blank(), \n",
    "        axis.text.y=element_blank(), \n",
    "        axis.ticks.y=element_blank()\n",
    "    )\n",
    "})\n",
    "do.call(cowplot::plot_grid, c(list(ncol = 3),p2))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2699713a-be9e-49aa-b5b8-8b5df7702db8",
   "metadata": {},
   "source": [
    "# Peak Calling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c854b3b3-c770-4565-8b84-374348cda094",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "proj_subset <- addGroupCoverages(ArchRProj = proj_subset, groupBy = \"pure_clusters\", maxReplicates = 16, maxCells = 2000, force = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb848e89-0ad6-447b-b2e8-e116721d6497",
   "metadata": {},
   "outputs": [],
   "source": [
    "pathToMacs2 <- findMacs2()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b384df52-35e6-4876-89ab-e783c0baec0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "proj_subset <- addReproduciblePeakSet(\n",
    "    ArchRProj = proj_subset, \n",
    "    groupBy = \"pure_clusters\", \n",
    "    pathToMacs2 = pathToMacs2, force = TRUE\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e75daca0-872a-49f2-8900-2ff809401c5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "proj_subset <- addPeakMatrix(proj_subset)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac12a877-e452-4e74-9aae-98235ec1421b",
   "metadata": {},
   "source": [
    "# Marker Peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "552e07c3-8550-47d7-87d0-0447f26272e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "markersPeaks <- getMarkerFeatures(\n",
    "    ArchRProj = proj_subset, \n",
    "    useMatrix = \"PeakMatrix\", \n",
    "    groupBy = \"pure_clusters\",\n",
    "  bias = c(\"TSSEnrichment\", \"log10(nFrags)\"),\n",
    "  testMethod = \"wilcoxon\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2bf6e149-cc63-4303-b88a-5d1a781cfe17",
   "metadata": {},
   "outputs": [],
   "source": [
    "heatmapPeaks <- markerHeatmap(\n",
    "  seMarker = markersPeaks, \n",
    "  cutOff = \"FDR <= 0.1 & Log2FC >= 0.5\",\n",
    "  transpose = TRUE\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "264c5949-9355-4224-80b6-b128c0537a6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 16, repr.plot.height = 8)\n",
    "draw(heatmapPeaks, heatmap_legend_side = \"bot\", annotation_legend_side = \"bot\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1936e8d5-4c57-4a94-a79b-1125409de90d",
   "metadata": {},
   "source": [
    "## Marker Peaks by Age"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d2746e44-4a08-4dad-8202-e8e5cc965441",
   "metadata": {},
   "outputs": [],
   "source": [
    "proj_subset$age_celltype <- paste0(proj_subset$pediatric_senior,\"_\",proj_subset$pure_clusters)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c84141a9-f1a6-4075-bf85-91304297a9ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "markersPeaks_age <- getMarkerFeatures(\n",
    "    ArchRProj = proj_subset, \n",
    "    useMatrix = \"PeakMatrix\", \n",
    "    groupBy = \"age_celltype\",\n",
    "  bias = c(\"TSSEnrichment\", \"log10(nFrags)\"),\n",
    "  testMethod = \"wilcoxon\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93c14888-6e14-4e02-a8b2-088a9542270b",
   "metadata": {},
   "outputs": [],
   "source": [
    "heatmapPeaks_age <- markerHeatmap(\n",
    "  seMarker = markersPeaks_age, \n",
    "  cutOff = \"FDR <= 0.1 & Log2FC >= 0.5\",\n",
    "  transpose = FALSE\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69c68a25-8da7-471c-b4ca-d33172dcaf3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 8, repr.plot.height = 16)\n",
    "draw(heatmapPeaks_age, heatmap_legend_side = \"bot\", annotation_legend_side = \"bot\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4b35399-13b3-486f-92de-7bd913227997",
   "metadata": {},
   "source": [
    "# Motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be2d235c-2618-444f-ba5a-e4ce594aa6af",
   "metadata": {},
   "outputs": [],
   "source": [
    "proj_subset <- addMotifAnnotations(ArchRProj = proj_subset, motifSet = \"JASPAR2020\", name = \"Motif\", force = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "389b0fc2-a20f-4194-ad80-7d20baa78b16",
   "metadata": {},
   "outputs": [],
   "source": [
    "enrichMotifs <- peakAnnoEnrichment(\n",
    "    seMarker = markersPeaks,\n",
    "    ArchRProj = proj_subset,\n",
    "    peakAnnotation = \"Motif\",\n",
    "    cutOff = \"FDR <= 0.1 & Log2FC >= 0.5\"\n",
    "  )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5eff401-228f-4e57-bfa0-cb6c267e0fc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "enrichMotifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e616de4f-e8a2-4547-9141-61206e4324a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "heatmapEM <- plotEnrichHeatmap(enrichMotifs, n = 10, transpose = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa6126f6-ac61-4688-850b-65df9aa387c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 16, repr.plot.height = 8)\n",
    "ComplexHeatmap::draw(heatmapEM, heatmap_legend_side = \"bot\", annotation_legend_side = \"bot\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b1de549-f22f-44bf-a588-f49e8591caff",
   "metadata": {},
   "source": [
    "# ChromVar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee16f118-d6d1-4dab-ac66-3bc88e455ce4",
   "metadata": {},
   "outputs": [],
   "source": [
    "proj_subset <- addBgdPeaks(proj_subset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85a29a05-5d47-4e2e-b598-f100acf6d2ba",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "proj_subset <- addDeviationsMatrix(\n",
    "  ArchRProj = proj_subset, \n",
    "  peakAnnotation = \"Motif\",\n",
    "  force = TRUE\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afe85f50-3657-49f9-a4d7-bfd7fa29164c",
   "metadata": {},
   "outputs": [],
   "source": [
    "motifs <- c('RELA','ATF7','TCFL5','KLF15','IRF4','SP1')\n",
    "markerMotifs <- getFeatures(proj_subset, select = paste(motifs, collapse=\"|\"), useMatrix = \"MotifMatrix\")\n",
    "markerMotifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "69b68b0e-fb2d-45b1-801a-4925ad34eaec",
   "metadata": {},
   "outputs": [],
   "source": [
    "markerMotifs <- grep(\"z:\", markerMotifs, value = TRUE)\n",
    "# markerMotifs <- markerMotifs[markerMotifs %ni% \"z:SREBF1_22\"]\n",
    "markerMotifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d630cba-0c86-474e-a8dd-53599f53235c",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width = 16, repr.plot.height = 8)\n",
    "p <- plotEmbedding(\n",
    "    ArchRProj = proj_subset, \n",
    "    colorBy = \"MotifMatrix\", \n",
    "    name = sort(markerMotifs), \n",
    "    embedding = \"UMAP\",\n",
    "    imputeWeights = getImputeWeights(proj_subset)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b8a05d5-a6f1-4ecd-b747-d3d1ff6fe5c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# p[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bbe88fc5-0c0b-4bc5-a567-6a364912d0ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 18, repr.plot.height = 18)\n",
    "p2 <- lapply(p, function(x){\n",
    "    x + guides(color = FALSE, fill = FALSE) + \n",
    "    theme_ArchR(baseSize = 6.5) +\n",
    "    theme(plot.margin = unit(c(0, 0, 0, 0), \"cm\")) +\n",
    "    theme(\n",
    "        axis.text.x=element_blank(), \n",
    "        axis.ticks.x=element_blank(), \n",
    "        axis.text.y=element_blank(), \n",
    "        axis.ticks.y=element_blank()\n",
    "    )\n",
    "})\n",
    "do.call(cowplot::plot_grid, c(list(ncol = 3),p2))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f5aec5a8-2580-479b-849f-06a7c91c3115",
   "metadata": {},
   "source": [
    "### Add WNN UMAP to ArchR Project"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5cd1954b-b259-4574-bd2a-889385146125",
   "metadata": {},
   "outputs": [],
   "source": [
    "addEmbedding <- function(ArchRProj= NULL, name= NULL, type = \"Generic\", dfEmbedding = NULL,model= NULL, modelName= NULL, embeddingParams= NULL){\n",
    "\n",
    "\tif(any(! rownames(dfEmbedding) %in% getCellNames(ArchRProj))){\n",
    "\t\tstop(\"All cell names for the embedding dataframe must be present in the ArchR  Project\")\n",
    "\t}\n",
    "\t\n",
    "\tif(!is.null(model)){\n",
    "\t\tdir <- getOutputDirectory(ArchRProj)\n",
    "\t\troute <- paste(dir,\"/Embeddings/\",modelName,\".rds\",sep = \"\")\n",
    "\n",
    "\t\tsaveRDS(model, file = route)\n",
    "\t}else{\n",
    "\n",
    "\t\troute = NA\n",
    "\n",
    "\t}\n",
    "\n",
    "\tcolnames(dfEmbedding) = paste(type,colnames(dfEmbedding),sep = \"#\")\n",
    "\t\n",
    "\tif(!is.null(embeddingParams)){\n",
    "\n",
    "\t\tArchRProj@embeddings[[name]] <- SimpleList(\n",
    "      \t\t\tdf = dfEmbedding, \n",
    "      \t\t\tparams = c(\n",
    "        \t\tembeddingParams,\n",
    "        \t\tdimsToUse = dimsToUse,\n",
    "        \t\tscaleDims = scaleDims,\n",
    "        \t\tcorCutOff = corCutOff,\n",
    "        \t\tnr=nr,\n",
    "        \t\tnc=nc,\n",
    "        \t\tuwotModel = route,\n",
    "        \t\testimateUMAP = estimateUMAP,\n",
    "        \t\tprojectID = projectDF))\n",
    "\t}else{\n",
    "\t\tArchRProj@embeddings[[name]] <- SimpleList(\n",
    "      \t\t\tdf = dfEmbedding)\n",
    "\t}\n",
    "\n",
    "\treturn(ArchRProj)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6f9644f-63c5-4df0-8547-918e63cd85db",
   "metadata": {},
   "outputs": [],
   "source": [
    "wnn_umap <- cd4_na@reductions$wnn.3.umap@cell.embeddings\n",
    "rownames(wnn_umap) <- paste(cd4_na$batch_id, \"-P1_\",cd4_na$pbmc_sample_id,\"#\",rownames(wnn_umap), sep = \"\")\n",
    "head(wnn_umap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ac023ba-a25f-4289-b2ad-681c3109767f",
   "metadata": {},
   "outputs": [],
   "source": [
    "table(rownames(wnn_umap) %in% proj_subset$cellNames)\n",
    "table(rownames(wnn_umap) == proj_subset$cellNames)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98edf5e3-bd9e-4e33-b809-26f2c6b13e9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "wnn_umap_ordered <- wnn_umap[match(proj_subset$cellNames, rownames(wnn_umap)),]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f99c55e6-5a94-4549-bc8d-46c3c7925adb",
   "metadata": {},
   "outputs": [],
   "source": [
    "table(rownames(wnn_umap_ordered) == proj_subset$cellNames)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64dbf2bd-906c-4926-8b6c-bcaaba45b4a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "proj_subset <- addEmbedding(ArchRProj = proj_subset, name = 'WNN_UMAP', dfEmbedding = wnn_umap_ordered)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f80a5c7-6910-4860-9ec7-15b3be5adca1",
   "metadata": {},
   "outputs": [],
   "source": [
    "motifs <- c('CREM','ATF7','TCFL5','KLF15','IRF8','SP1')\n",
    "markerMotifs <- getFeatures(proj_subset, select = paste(motifs, collapse=\"|\"), useMatrix = \"MotifMatrix\")\n",
    "markerMotifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb47b55c-63d5-4732-b702-e2a6b6b76a19",
   "metadata": {},
   "outputs": [],
   "source": [
    "markerMotifs <- grep(\"z:\", markerMotifs, value = TRUE)\n",
    "# markerMotifs <- markerMotifs[markerMotifs %ni% \"z:SREBF1_22\"]\n",
    "markerMotifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afa225ee-b434-4f90-9753-6affa54d1bd7",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width = 16, repr.plot.height = 8)\n",
    "p <- plotEmbedding(\n",
    "    ArchRProj = proj_subset, \n",
    "    colorBy = \"MotifMatrix\", \n",
    "    name = sort(markerMotifs), \n",
    "    embedding = \"WNN_UMAP\",\n",
    "    imputeWeights = getImputeWeights(proj_subset)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "45231fa8-ab46-4c1a-bceb-4088e272c751",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 18, repr.plot.height = 12)\n",
    "p2 <- lapply(p, function(x){\n",
    "    x + guides(color = FALSE, fill = FALSE) + \n",
    "    theme_ArchR(baseSize = 6.5) +\n",
    "    theme(plot.margin = unit(c(0, 0, 0, 0), \"cm\")) +\n",
    "    theme(\n",
    "        axis.text.x=element_blank(), \n",
    "        axis.ticks.x=element_blank(), \n",
    "        axis.text.y=element_blank(), \n",
    "        axis.ticks.y=element_blank()\n",
    "    )\n",
    "})\n",
    "do.call(cowplot::plot_grid, c(list(ncol = 3),p2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc4bc277-4f3a-4e79-8ca5-7de65e71ed42",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Open a pdf file\n",
    "pdf(\"plots/cd4na_chromvar_umaps.pdf\", width = 18, height = 12) \n",
    "# 2. Create a plot\n",
    "p2 <- lapply(p, function(x){\n",
    "    x + guides(color = FALSE, fill = FALSE) + \n",
    "    theme_ArchR(baseSize = 6.5) +\n",
    "    theme(plot.margin = unit(c(0, 0, 0, 0), \"cm\")) +\n",
    "    theme(\n",
    "        axis.text.x=element_blank(), \n",
    "        axis.ticks.x=element_blank(), \n",
    "        axis.text.y=element_blank(), \n",
    "        axis.ticks.y=element_blank()\n",
    "    )\n",
    "})\n",
    "do.call(cowplot::plot_grid, c(list(ncol = 3),p2))\n",
    "# Close the pdf file\n",
    "dev.off() "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "483c7e70-43d3-437f-be14-8c70f0cfd102",
   "metadata": {},
   "source": [
    "# Save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cecd57f-f5e3-441d-8db6-6f0b1375e4a1",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "saveArchRProject(ArchRProj = proj_subset, outputDirectory = 'ATAC/', load = F)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b1bcd5a-4ad7-45d9-a2cf-45d2ee8d17ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "proj_subset <- loadArchRProject(path = 'ATAC/')\n",
    "proj_subset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27a2975a-b3d9-4263-a608-697b2ab24d9a",
   "metadata": {},
   "source": [
    "# Tracks"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2bccc8f3-0993-49a3-8059-e9a7795cb3a3",
   "metadata": {},
   "source": [
    "## Plotting Functions by MP and LO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "195fdc83-e1a5-4040-880b-ada4903a0055",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "### These functions are meant for exporting ATAC data from an ArchR Project, whether into plots or various Bed/BedGraph/BigWig files\n",
    "\n",
    "\n",
    "library(ArchR)\n",
    "library(dplyr)\n",
    "library(tidyverse)\n",
    "library(plyranges)\n",
    "library(TxDb.Hsapiens.UCSC.hg38.refGene)\n",
    "library(doParallel)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "#########################  Section 1: Exporting ATAC data into Coverage/Insertions file at single basepair resolution ----\n",
    "\n",
    "\n",
    "## This will allow for cell-count normalized insertion and coverage file creation.\n",
    "## Export options\n",
    "\n",
    "### Here's the pipeline:\n",
    "\n",
    "### Step 1: Extract fragments by populations from an ArchR Project: getPopFrags()\n",
    "### Step 2: Export to the file type that you desire\n",
    "################  FragsToCoverageFiles: Export fragments to BigWig/BedGraph files that contain coverage counts per 1000 cells.\n",
    "################  FragsToInsertionFiles: Export fragments to BigWig/BedGraph files that contain insertion counts per 1000 cells.\n",
    "\n",
    "\n",
    "#' Extract fragments by populations from an ArchR Project\n",
    "#'\n",
    "#' \\code{getPopFrags} returns a list of fragments per cell subset as a GRanges.\n",
    "#'\n",
    "#' @param ArchRProj The ArchR Project.\n",
    "#' @param metaColumn The name of metadata column that contains the populations\n",
    "#'   of cells you want to merge and export.\n",
    "#' @param cellSubsets Default is 'ALL'. If you want to export only some groups,\n",
    "#'   then give it a list of group names. This needs to be unique - no duplicated\n",
    "#'   names. This list of group names must be identical to names that appear in\n",
    "#'   the metadata column of the ArchR Project (e.g. metaColumn).\n",
    "#' @param region Optional parameter. Set this if you only want to extract\n",
    "#'   fragments from particular regions of the genome. Format should be as a\n",
    "#'   string (e.g. 'chr1:1000-2000'), or a GRanges object.\n",
    "#' @param numCores Number of cores to use.\n",
    "#' @param sampleSpecific Set to TRUE to further subset cells by sample\n",
    "#' @param NormMethod Normalization method. Can be either \"nFrags\",\"nCells\", or\n",
    "#'   \"Median\".\n",
    "#' @param blackList Blacklisted region to filter out. Default is to not filter\n",
    "#'   out anything (i.e. NULL). Input should be provided as a GRanges object. Any\n",
    "#'   fragments with more than a certain overlap will be thrown out.\n",
    "#' @param overlapList The minimum overlap necessary for a fragment marked as\n",
    "#'   overlapping with the blacklist region and thus thrown out.\n",
    "#'\n",
    "#' @return A list of GRanges containing fragments. Each GRanges corresponds to a\n",
    "#'   population defined by cellSubsets (and sample, if\n",
    "#'   \\code{sampleSpecific=TRUE})\n",
    "#'\n",
    "getPopFrags <- function(ArchRProj, metaColumn, cellSubsets = \"ALL\", region = NULL, numCores = 1, sampleSpecific = TRUE,\n",
    "                        NormMethod = \"nfrags\", blackList = NULL, overlapList = 50) {\n",
    "  # Extract metadata\n",
    "  metadf <- getCellColData(ArchRProj)\n",
    "\n",
    "  if (!any(colnames(metadf) %in% metaColumn)) {\n",
    "    stop(paste(\n",
    "      str_interp(\"Provided metaColumn ${metaColumn} does not exist in the cellColData of your ArchRProj.\"),\n",
    "      str_interp(\"Available columns are: ${colnames(metadf)}\")\n",
    "    ))\n",
    "  }\n",
    "\n",
    "  # Get cell counts for all cell populations\n",
    "  tmp <- metadf[, metaColumn]\n",
    "  cellCounts <- table(tmp[!is.na(tmp)])\n",
    "\n",
    "  if (all(cellSubsets == \"ALL\")) {\n",
    "    cellPopulations <- names(cellCounts)\n",
    "  } else {\n",
    "\n",
    "    # Take our cell populations from given cellSubsets\n",
    "    cellPopulations <- unique(cellSubsets[!is.na(cellSubsets)][order(cellSubsets)])\n",
    "\n",
    "    # Filter cellCounts to selected cell subsets (populations)\n",
    "    cellCounts <- cellCounts[cellPopulations]\n",
    "    if (any(is.na(cellCounts))) {\n",
    "      stop(paste(\n",
    "        \"Some cell populations have NA cell counts.\",\n",
    "        \"Please verify that all given cellSubsets exist for the given metaColumn.\",\n",
    "        str_interp(\"cellSubsets with NA cell counts: ${cellPopulations[is.na(names(cellCounts))]}\")\n",
    "      ))\n",
    "    }\n",
    "  }\n",
    "\n",
    "  if (length(cellCounts) == 0) {\n",
    "    stop(\"No cells were found for the given cellSubsets and/or metaColumn.\")\n",
    "  }\n",
    "\n",
    "\n",
    "  #### Up to here, the above is only specific to sampleSpecific=False\n",
    "  allArrows <- getArrowFiles(ArchRProj)\n",
    "  # Get sample names from our populations of interest\n",
    "  samplesToExtract <- paste(\n",
    "    unique(metadf$Sample[which(metadf[, metaColumn] %in% cellPopulations)]),\n",
    "    collapse = \"|\"\n",
    "  )\n",
    "  # Get arrows containing the samples we want\n",
    "  arrows <- allArrows[grepl(samplesToExtract, allArrows)]\n",
    "\n",
    "  if (length(arrows) == 0) {\n",
    "    stop(paste(\n",
    "      \"Found no arrows containing samples from your selected cell subset(s).\",\n",
    "      \"Check cellSubsets or the input ArchR Project.\"\n",
    "    ))\n",
    "  }\n",
    "\n",
    "  if (is.null(region)) {\n",
    "\n",
    "    # Extract fragments from all available regions\n",
    "    fragsList <- mclapply(seq_along(arrows), function(x) {\n",
    "      getFragmentsFromArrow(\n",
    "        ArrowFile = arrows[x],\n",
    "        cellNames = NULL,\n",
    "        verbose = FALSE\n",
    "      )\n",
    "    }, mc.cores = numCores)\n",
    "  } else {\n",
    "\n",
    "    # Extract fragments from a given region only\n",
    "    # Validate region and interpret as string or GRanges\n",
    "    if (validRegionString(region) & tolower(NormMethod) == \"raw\") {\n",
    "      regionGRanges <- StringsToGRanges(region)\n",
    "    } else if (class(region)[1] == \"GRanges\" & tolower(NormMethod) == \"raw\") {\n",
    "      regionGRanges <- region\n",
    "    } else if (tolower(NormMethod) != \"raw\") {\n",
    "      stop(paste(\n",
    "        \"Wrong NormMethod set. Please set NormMethod = 'Raw'\",\n",
    "        \"if you wish to extract fragments from a specific region.\",\n",
    "        \"\\n Alternatively, you can extract the entire genome and then subset.\"\n",
    "      ))\n",
    "    } else {\n",
    "      stop(paste(\n",
    "        \"Invalid region input.\",\n",
    "        \"Region must either be a string matching the format\",\n",
    "        \"'seqname:start-end', or a GRanges object.\"\n",
    "      ))\n",
    "    }\n",
    "\n",
    "    chrom <- regionGRanges %>%\n",
    "      as.data.frame() %>%\n",
    "      dplyr::select(seqnames)\n",
    "    fragsList <- mclapply(seq_along(arrows), function(x) {\n",
    "      fragsGRanges <- getFragmentsFromArrow(\n",
    "        ArrowFile = arrows[x],\n",
    "        cellNames = NULL,\n",
    "        chr = as.character(chrom[, 1]),\n",
    "        verbose = FALSE\n",
    "      ) %>% plyranges::join_overlap_intersect(regionGRanges)\n",
    "\n",
    "      # Filter according to provided blacklist\n",
    "      if (is.null(blackList)) {\n",
    "        fragsGRanges\n",
    "      } else if (class(blackList)[1] == \"GRanges\") {\n",
    "        fragsGRanges %>% plyranges::filter_by_non_overlaps(blackList, minoverlap = overlapList)\n",
    "      } else {\n",
    "        stop(\"Error: Wrong format for blackList region. Please provide GRanges\")\n",
    "      }\n",
    "    }, mc.cores = numCores)\n",
    "  }\n",
    "\n",
    "  # From scMACS - sorts cell barcodes by population\n",
    "  barcodesByCellPop <- lapply(cellPopulations, function(x) {\n",
    "    row.names(metadf)[which(metadf[, metaColumn] == x)]\n",
    "  })\n",
    "\n",
    "  # Add normalization factor.\n",
    "  if (tolower(NormMethod) == \"raw\") {\n",
    "    names(barcodesByCellPop) <- paste(gsub(\"#\", \"_\", gsub(\" |_\", \"_\", cellPopulations)), 1, sep = \"__\")\n",
    "  } else if (tolower(NormMethod) == \"ncells\") {\n",
    "    names(barcodesByCellPop) <- paste(gsub(\"#\", \"_\", gsub(\" |_\", \"_\", cellPopulations)), cellCounts / 1000, sep = \"__\")\n",
    "  } else if (tolower(NormMethod) == \"nfrags\") {\n",
    "    # Calculate the total nFrags for our cell populations\n",
    "    nFragsNorm <- as.data.frame(metadf[, c(metaColumn, \"nFrags\")]) %>%\n",
    "      dplyr::filter(!is.na(get(metaColumn))) %>%\n",
    "      dplyr::filter(get(metaColumn) %in% cellPopulations) %>%\n",
    "      dplyr::group_by(get(metaColumn)) %>%\n",
    "      dplyr::summarize(nFrags = sum(nFrags))\n",
    "\n",
    "    # Verify expected cellPopulations have corresponding nFrags\n",
    "    if (!all(nFragsNorm[, 1] == cellPopulations)) {\n",
    "      print(nFragsNorm[, 1])\n",
    "      stop(\"Names of nFrags don't match the cell populations.\")\n",
    "    }\n",
    "    names(barcodesByCellPop) <- paste(gsub(\"#\", \"_\", gsub(\" |_\", \"_\", cellPopulations)), nFragsNorm$nFrags / 10^6, sep = \"__\")\n",
    "  } else if (tolower(NormMethod) == \"median\") {\n",
    "    names(barcodesByCellPop) <- paste(gsub(\"#\", \"_\", gsub(\" |_\", \"_\", cellPopulations)), \"Median\", sep = \"__\")\n",
    "  } else if (tolower(NormMethod) == \"medmax\") {\n",
    "    names(barcodesByCellPop) <- paste(gsub(\"#\", \"_\", gsub(\" |_\", \"_\", cellPopulations)), \"MedianMax\", sep = \"__\")\n",
    "  } else if (tolower(NormMethod) == \"TotalSampleByNCells\") {\n",
    "    # This calculates total fragments per sample.\n",
    "    totalsampleFrags <- unlist(fragsList, length)\n",
    "    names(barcodesByCellPop) <- paste(gsub(\"#\", \"_\", gsub(\" |_\", \"_\", cellPopulations)), totalsampleFrags / 10^6 * cellCounts / 1000, sep = \"__\")\n",
    "  } else {\n",
    "    stop(\"Error: Incorrect NormMethod given.\")\n",
    "  }\n",
    "\n",
    "  # From scMACS - Function to sort fragments by populations based on cell barcode lists\n",
    "  subset_Frag <- function(cellNames, fragsGRanges) {\n",
    "    fragsTable <- as.data.table(fragsGRanges)\n",
    "    idx <- which(fragsTable$RG %in% cellNames)\n",
    "    fragsGRanges[idx]\n",
    "  }\n",
    "\n",
    "  ## Identify which subset of arrows the population can be found it.\n",
    "  ## Speeds up sample-specific population extraction\n",
    "  fragsListIndex <- lapply(cellPopulations, function(x) {\n",
    "    names(arrows) %in% unique(metadf$Sample[which(metadf[, metaColumn] == x)])\n",
    "  })\n",
    "\n",
    "  # Sort fragments into a list by cell population\n",
    "\n",
    "  popFrags <- lapply(seq_along(barcodesByCellPop), function(x) {\n",
    "    print(paste(\"Sorting \", names(barcodesByCellPop)[x], sep = \"\"))\n",
    "    if (sum(fragsListIndex[[x]]) > 1) {\n",
    "      tmp <- mclapply(which(fragsListIndex[[x]]), function(y) {\n",
    "        subset_Frag(barcodesByCellPop[[x]], fragsList[[y]])\n",
    "      }, mc.cores = 20)\n",
    "    } else {\n",
    "      tmp <- list(subset_Frag(\n",
    "        barcodesByCellPop[[x]],\n",
    "        fragsList[[which(fragsListIndex[[x]])]]\n",
    "      ))\n",
    "    }\n",
    "\n",
    "    # For this population, get sample-specific normalization factors\n",
    "    # and rename the fragment GRanges\n",
    "    if (sampleSpecific) {\n",
    "      subSet_tmp <- gsub(\"__.*\", \"\", names(barcodesByCellPop)[x])\n",
    "\n",
    "      if (tolower(NormMethod) == \"ncells\") {\n",
    "        tmp_cellCount <- unlist(lapply(tmp, function(y) {\n",
    "          length(unique(y$RG))\n",
    "        }))\n",
    "\n",
    "        names(tmp) <- paste(\n",
    "          subSet_tmp,\n",
    "          \"#\",\n",
    "          names(arrows)[unlist(fragsListIndex[[x]])],\n",
    "          \"__\",\n",
    "          tmp_cellCount / 1000,\n",
    "          sep = \"\"\n",
    "        )\n",
    "      } else if (tolower(NormMethod) == \"nfrags\") {\n",
    "        names(tmp) <- paste(\n",
    "          subSet_tmp,\n",
    "          \"#\",\n",
    "          names(arrows)[unlist(fragsListIndex[[x]])],\n",
    "          \"__\",\n",
    "          unlist(lapply(tmp, length)) / 10^6,\n",
    "          sep = \"\"\n",
    "        )\n",
    "      } else {\n",
    "        subSet_tmp <- gsub(\"__.*\", \"\", names(barcodesByCellPop)[x])\n",
    "        norm_tmp <- gsub(\".*__\", \"\", names(barcodesByCellPop)[x])\n",
    "\n",
    "        names(tmp) <- paste(\n",
    "          subSet_tmp,\n",
    "          \"#\",\n",
    "          names(arrows)[unlist(fragsListIndex[[x]])], \"__\", norm_tmp,\n",
    "          sep = \"\"\n",
    "        )\n",
    "      }\n",
    "\n",
    "      tmp\n",
    "    } else {\n",
    "      stack(as(tmp, \"GRangesList\"))\n",
    "    }\n",
    "  })\n",
    "\n",
    "  if (sampleSpecific) {\n",
    "    popFrags <- unlist(popFrags, recursive = FALSE)\n",
    "  } else {\n",
    "    names(popFrags) <- names(barcodesByCellPop)\n",
    "  }\n",
    "  return(popFrags)\n",
    "}\n",
    "\n",
    "\n",
    "###### This function take the output from getPopFrags and figures out how to normalize before exporting to the desired file formats.\n",
    "## popFrags = output from getPopFrags, as a list for the sake of parallelization. The name for each population is populatioName_Norm\n",
    "## metadata = ArchR metadata\n",
    "## metaColumn = column in metadata with popFrags\n",
    "## outDir = the directory to output the resultant fragment files. Optional - default is the current working directory.\n",
    "FragsToCoverageFiles <- function(popFrags, files = \"BigWig\", genome = \"hg38\", sampleAvg = TRUE, fname = NULL, outDir = NULL, numCores = 1) {\n",
    "\n",
    "  # Extract cell counts for each group\n",
    "  groups <- gsub(\"__.*\", \"\", names(popFrags))\n",
    "  allNorm <- gsub(\".*__\", \"\", names(popFrags))\n",
    "  names(allNorm) <- groups\n",
    "\n",
    "  # Make bed file names for each group, either with names from popFrags or from the fname list.\n",
    "  if (is.null(fname) & !sampleAvg) {\n",
    "    bedFiles <- paste(gsub(\"-\", \"_\", gsub(\"/\", \"_\", gsub(\" \", \"_\", groups, fixed = TRUE))), \".bed\", sep = \"\")\n",
    "  } else if (length(fname) == length(popFrags) & !sampleAvg) {\n",
    "\n",
    "    # Check if fname is the same length as the popFrags and we're not doing sample Avg\n",
    "    bedFiles <- paste(fname, \".bed\", sep = \"\")\n",
    "  } else if (is.null(fname) & sampleAvg & all(grepl(\"#\", names(popFrags)))) {\n",
    "\n",
    "    # If fname is null and sampleAvg is true, then the popFrags needs to be formated for sample-specific coverage.\n",
    "\n",
    "    uniqueGroups <- unique(gsub(\"#.*\", \"#\", names(popFrags)))\n",
    "    bedFiles <- paste(gsub(\"-\", \"_\", gsub(\"/\", \"_\", gsub(\" \", \"_\", groups, fixed = TRUE))), \".bed\", sep = \"\")\n",
    "  } else if (length(fname) == length(unique(gsub(\"#.*\", \"#\", names(popFrags)))) & sampleAvg & all(grepl(\"#\", names(popFrags)))) {\n",
    "\n",
    "    # If we want to rename the sample average, then fname needs to be the same length as the number of unique groups.\n",
    "\n",
    "    uniqueGroups <- unique(gsub(\"#.*\", \"#\", names(popFrags)))\n",
    "    bedFiles <- paste(fname, \".bed\", sep = \"\")\n",
    "  } else if (sampleAvg == TRUE & !all(grepl(\"#\", names(popFrags)))) {\n",
    "    stop(\"Error: You asked for exporting sample averages for all cell groups, but didn't provide sample-specific fragment lists.\n",
    "\t\tPlease re-run getPopFrags with sampleSpecific = TRUE\")\n",
    "  } else {\n",
    "    stop(\"Error: fname parameter is not the same length as the number of populations you want to export.\")\n",
    "  }\n",
    "  # Prepend output directory (if given) to file names\n",
    "  if (!is.null(outDir)) {\n",
    "    dir.create(outDir, recursive = TRUE)\n",
    "    bedFiles <- file.path(outDir, bedFiles)\n",
    "  }\n",
    "\n",
    "  # Test export options for bed and bam files\n",
    "  if (all(tolower(files) == \"all\" & !sampleAvg)) {\n",
    "\n",
    "    # Export to BAM\n",
    "    writeBAMHelper(popFrags, bedFiles, genome, numCores)\n",
    "    # Export data from GRanges to bed\n",
    "    mclapply(seq_along(popFrags), function(x) {\n",
    "      # Export to bed\n",
    "      plyranges::write_bed(popFrags[[x]], bedFiles[x])\n",
    "    }, mc.cores = numCores)\n",
    "  } else if (any(tolower(files) == \"bed\" & !sampleAvg)) {\n",
    "\n",
    "    # Export data from GRanges to bed files, and end the function.\n",
    "    mclapply(seq_along(popFrags), function(x) {\n",
    "      plyranges::write_bed(popFrags[[x]], bedFiles[x])\n",
    "    }, mc.cores = numCores)\n",
    "    return()\n",
    "  } else if (any(tolower(files) == \"bam\" & !sampleAvg)) {\n",
    "    # Export to BAM, and end the function\n",
    "    writeBAMHelper(popFrags, bedFiles, genome, numCores)\n",
    "    return()\n",
    "  }\n",
    "\n",
    "  ### Let's generate coverage files within R,\n",
    "\n",
    "\n",
    "  # If you want to take the sample average across a group\n",
    "\n",
    "  if (sampleAvg == TRUE & all(grepl(\"#\", names(popFrags)))) {\n",
    "    mclapply(seq_along(uniqueGroups), function(x) {\n",
    "      print(paste(\"Summarizing\", uniqueGroups[x], sep = \"\"))\n",
    "\n",
    "      # Iterate along all samples for this population, calculating the coverage, and then turn the GRanges into a data.table\n",
    "      # After which, we bind those data.tables together, and turn it back into a GRanges object.\n",
    "      # This might seem ineffecient, but computationally, it's faster and less prone to error as plyranges is actively being developed.\n",
    "\n",
    "      if (grepl(\"Median|Max|MedMax\", allNorm[x])) {\n",
    "\n",
    "        # Iterate along all the samples of this cell group.\n",
    "        tmp1 <- lapply(seq_along(which(grepl(uniqueGroups[x], names(popFrags)))), function(y) {\n",
    "          tmp_Cov <- plyranges::compute_coverage(popFrags[[y]])\n",
    "\n",
    "          medCov <- tmp_Cov %>%\n",
    "            plyranges::filter(score != 0) %>%\n",
    "            plyranges::tile_ranges(., width = 1)\n",
    "\n",
    "          maxScore <- max(tmp_Cov$score[medCov$partition])\n",
    "          medScore <- median(tmp_Cov$score[medCov$partition])\n",
    "          medMax <- max(tmp_Cov$score[medCov$partition] / medScore)\n",
    "\n",
    "          tmp_Cov %>%\n",
    "            plyranges::mutate(Norm = rep(NormVal, n = plyranges::n())) %>%\n",
    "            plyranges::mutate(score = case_when(\n",
    "              grepl(\"Median\", Norm) ~ score / medScore,\n",
    "              grepl(\"Max\", Norm) ~ score / maxScore,\n",
    "              grepl(\"MedMax\", Norm) ~ score / medMax,\n",
    "              TRUE ~ score / as.numeric(Norm)\n",
    "            )) %>% # Redundant/safety measure. Should not be triggered.\n",
    "            as.data.table()\n",
    "\n",
    "\n",
    "          tmp_Cov\n",
    "        }) %>%\n",
    "          rbindlist() %>%\n",
    "          makeGRangesFromDataFrame(., keep.extra.columns = TRUE)\n",
    "      } else {\n",
    "        tmp1 <- lapply(seq_along(which(grepl(uniqueGroups[x], names(popFrags)))), function(y) {\n",
    "\n",
    "          # Compute coverage for each sample, and normalize by the normalization factor.\n",
    "\n",
    "          NormVal <- unique(gsub(\".*__\", \"\", names(popFrags)[y]))\n",
    "          plyranges::compute_coverage(popFrags[[y]]) %>%\n",
    "            plyranges::mutate(name = names(popFrags)[y]) %>%\n",
    "            plyranges::mutate(score = score / as.numeric(NormVal)) %>%\n",
    "            as.data.table()\n",
    "        }) %>%\n",
    "          rbindlist() %>%\n",
    "          makeGRangesFromDataFrame(., keep.extra.columns = TRUE)\n",
    "      }\n",
    "\n",
    "      sampleNum <- sum(grepl(uniqueGroups[x], names(popFrags)))\n",
    "      popCov <- tmp1 %>% disjoin_ranges(score = sum(score) / sampleNum)\n",
    "      # return(popCov)\n",
    "      exportCov(\n",
    "        CovObject = popCov, files = files,\n",
    "        fileName = gsub(\"#\", \"\", uniqueGroups[x])\n",
    "      )\n",
    "    }, mc.cores = numCores)\n",
    "  } else {\n",
    "    mclapply(seq_along(popFrags), function(x) {\n",
    "      print(paste(\"Summarizing \", gsub(\"__.*\", \"\", bedFiles[x]), sep = \"\"))\n",
    "\n",
    "      tmp1 <- plyranges::compute_coverage(popFrags[[x]])\n",
    "\n",
    "      if (grepl(\"Median|Max|MedMax\", allNorm[x])) {\n",
    "        medCov <- tmp1 %>%\n",
    "          plyranges::filter(score != 0) %>%\n",
    "          plyranges::tile_ranges(., width = 1)\n",
    "\n",
    "        maxScore <- max(tmp1$score[medCov$partition])\n",
    "        medScore <- median(tmp1$score[medCov$partition])\n",
    "        medMax <- max(tmp1$score[medCov$partition] / medScore)\n",
    "\n",
    "        popCov <- tmp1 %>%\n",
    "          plyranges::mutate(Norm = rep(allNorm[[x]], n = plyranges::n())) %>%\n",
    "          plyranges::mutate(score = case_when(\n",
    "            grepl(\"Median\", Norm) ~ score / medScore,\n",
    "            grepl(\"Max\", Norm) ~ score / maxScore,\n",
    "            grepl(\"MedMax\", Norm) ~ score / medMax,\n",
    "            TRUE ~ score / as.numeric(Norm)\n",
    "          )) %>%\n",
    "          plyranges::mutate(score = ifelse(grepl(\"MedMax\", Norm), score / medMax))\n",
    "      } else {\n",
    "        popCov <- tmp1 %>%\n",
    "          plyranges::mutate(Norm = rep(allNorm[[x]], n = plyranges::n())) %>%\n",
    "          plyranges::mutate(score = score / as.numeric(Norm))\n",
    "      }\n",
    "\n",
    "      exportCov(\n",
    "        CovObject = popCov, files = files,\n",
    "        fileName = gsub(\".bed\", \"\", bedFiles[x])\n",
    "      )\n",
    "    }, mc.cores = numCores)\n",
    "  }\n",
    "}\n",
    "\n",
    "\n",
    "### Similar function to above, except will export insertion counts.\n",
    "FragsToInsertionFiles <- function(popFrags, files = \"bw\", genome = \"hg38\", fname = NULL, outDir = NULL, numCores = 1) {\n",
    "\n",
    "  # Double check inputs: Are the files that we need actually present?\n",
    "\n",
    "  # Extract cell counts for each group\n",
    "  groups <- gsub(\"__.*\", \"\", names(popFrags))\n",
    "  allNorm <- gsub(\".*__\", \"\", names(popFrags))\n",
    "  names(allNorm) <- groups\n",
    "\n",
    "  ## Let's extract Insertion sites\n",
    "  popIns <- mclapply(seq_along(popFrags), function(x) {\n",
    "    cutstart <- GRanges(seqnames = as(seqnames(popFrags[[x]]), \"vector\"), ranges = IRanges(start = start(popFrags[[x]]), width = 1), strand = \"*\")\n",
    "    cutend <- GRanges(seqnames = as(seqnames(popFrags[[x]]), \"vector\"), ranges = IRanges(start = end(popFrags[[x]]), width = 1), strand = \"*\")\n",
    "    plyranges::bind_ranges(cutstart, cutend)\n",
    "  }, mc.cores = numCores)\n",
    "  names(popIns) <- names(popFrags)\n",
    "\n",
    "\n",
    "  # Make bed file names for each group, either with names from popFrags or from the fname list.\n",
    "  if (is.null(fname)) {\n",
    "    bedFiles2 <- paste(gsub(\"-\", \"_\", gsub(\"/\", \"_\", gsub(\" \", \"_\", groups, fixed = TRUE))), \"_insertions.bed\", sep = \"\")\n",
    "  } else if (length(fname) == length(popFrags)) {\n",
    "    bedFiles2 <- paste(fname, \"_insertions.bed\", sep = \"\")\n",
    "  } else {\n",
    "    stop(\"Error: fname parameter is not the same length as popFrags\")\n",
    "  }\n",
    "  # Prepend output directory (if given) to file names\n",
    "  if (!is.null(outDir)) {\n",
    "    dir.create(outDir, recursive = TRUE)\n",
    "    bedFiles2 <- file.path(outDir, bedFiles2)\n",
    "  }\n",
    "\n",
    "  if (all(tolower(files) == \"all\")) {\n",
    "\n",
    "\n",
    "    # Export to BAM\n",
    "    writeBAMHelper(popIns, bedFiles2, genome, numCores)\n",
    "    # Export data from GRanges to bed files\n",
    "    # Then move onto export the rest\n",
    "    mclapply(seq_along(popIns), function(x) {\n",
    "      plyranges::write_bed(popIns[[x]], bedFiles2[x])\n",
    "    }, mc.cores = numCores)\n",
    "  } else if (any(tolower(files) == \"bed\")) {\n",
    "\n",
    "    # Export data from GRanges to bed files, and end the function.\n",
    "    mclapply(seq_along(popIns), function(x) {\n",
    "      plyranges::write_bed(popIns[[x]], bedFiles2[x])\n",
    "    }, mc.cores = numCores)\n",
    "    return()\n",
    "  } else if (any(tolower(files) == \"bam\")) {\n",
    "\n",
    "    # Export to BAM, and end the function\n",
    "    writeBAMHelper(popIns, bedFiles2, genome, numCores)\n",
    "    return()\n",
    "  }\n",
    "\n",
    "\n",
    "  # Export to bigwigs or bed graphs.\n",
    "\n",
    "  # If you want to take the sample average across a group\n",
    "\n",
    "  if (sampleAvg == TRUE & all(grepl(\"#\", names(popIns)))) {\n",
    "    uniqueGroups <- unique(gsub(\"#.*\", \"#\", names(popIns)))\n",
    "\n",
    "\n",
    "    mclapply(seq_along(uniqueGroups), function(x) {\n",
    "      print(paste(\"Summarizing\", uniqueGroups[x], sep = \"\"))\n",
    "\n",
    "      # Iterate along all samples for this population, calculating the coverage, and then turn the GRanges into a data.table\n",
    "      # After which, we bind those data.tables together, and turn it back into a GRanges object.\n",
    "      # This might seem ineffecient, but computationally, it's faster and less prone to error as plyranges is actively being developed.\n",
    "\n",
    "      if (grepl(\"Median|Max|MedMax\", allNorm[x])) {\n",
    "\n",
    "        # Iterate along all the samples of this cell group.\n",
    "        tmp1 <- lapply(seq_along(which(grepl(uniqueGroups[x], names(popIns)))), function(y) {\n",
    "          tmp_Cov <- plyranges::compute_coverage(popIns[[y]])\n",
    "\n",
    "          medCov <- tmp_Cov %>%\n",
    "            plyranges::filter(score != 0) %>%\n",
    "            plyranges::tile_ranges(., width = 1)\n",
    "\n",
    "          maxScore <- max(tmp_Cov$score[medCov$partition])\n",
    "          medScore <- median(tmp_Cov$score[medCov$partition])\n",
    "          medMax <- max(tmp_Cov$score[medCov$partition] / medScore)\n",
    "\n",
    "          tmp_Cov %>%\n",
    "            plyranges::mutate(Norm = rep(NormVal, n = plyranges::n())) %>%\n",
    "            plyranges::mutate(score = case_when(\n",
    "              grepl(\"Median\", Norm) ~ score / medScore,\n",
    "              grepl(\"Max\", Norm) ~ score / maxScore,\n",
    "              grepl(\"MedMax\", Norm) ~ score / medMax,\n",
    "              TRUE ~ score / as.numeric(Norm)\n",
    "            )) %>% # Redundant/safety measure. Should not be triggered.\n",
    "            as.data.table()\n",
    "\n",
    "\n",
    "          tmp_Cov\n",
    "        }) %>%\n",
    "          rbindlist() %>%\n",
    "          makeGRangesFromDataFrame(., keep.extra.columns = TRUE)\n",
    "      } else {\n",
    "        tmp1 <- lapply(seq_along(which(grepl(uniqueGroups[x], names(popIns)))), function(y) {\n",
    "\n",
    "          # Compute coverage for each sample, and normalize by the normalization factor.\n",
    "\n",
    "          NormVal <- unique(gsub(\".*__\", \"\", names(popIns)[y]))\n",
    "          plyranges::compute_coverage(popIns[[y]]) %>%\n",
    "            plyranges::mutate(name = names(popIns)[y]) %>%\n",
    "            plyranges::mutate(score = score / as.numeric(NormVal)) %>%\n",
    "            as.data.table()\n",
    "        }) %>%\n",
    "          rbindlist() %>%\n",
    "          makeGRangesFromDataFrame(., keep.extra.columns = TRUE)\n",
    "      }\n",
    "\n",
    "      popCov <- tmp1 %>% disjoin_ranges(score = sum(score) / sum(grepl(uniqueGroups[x], names(popIns))))\n",
    "\n",
    "      exportCov(\n",
    "        CovObject = popCov, files = files,\n",
    "        fileName = gsub(\"#\", \"\", uniqueGroups[x])\n",
    "      )\n",
    "    }, mc.cores = numCores)\n",
    "  } else if (sampleAvg == TRUE) {\n",
    "    stop(\"Error: You asked for exporting sample averages for all cell groups, but didn't provide sample-specific fragment lists. Please re-run getPopFrags with sampleSpecific = TRUE\")\n",
    "  } else {\n",
    "    mclapply(seq_along(popIns), function(x) {\n",
    "      print(paste(\"Exporting \", gsub(\".bed\", \"\", bedFiles2[x]), sep = \"\"))\n",
    "      tmp1 <- plyranges::compute_coverage(popIns[[x]])\n",
    "\n",
    "      if (grepl(\"Median|Max|MedMax\", allNorm[x])) {\n",
    "        medCov <- tmp1 %>%\n",
    "          plyranges::filter(score != 0) %>%\n",
    "          plyranges::tile_ranges(., width = 1)\n",
    "\n",
    "        maxScore <- max(tmp1$score[medCov$partition])\n",
    "        medScore <- median(tmp1$score[medCov$partition])\n",
    "        medMax <- max(tmp1$score[medCov$partition] / medScore)\n",
    "\n",
    "        popCov <- tmp1 %>%\n",
    "          plyranges::mutate(Norm = rep(allNorm[[x]], n = plyranges::n())) %>%\n",
    "          plyranges::mutate(score = case_when(\n",
    "            grepl(\"Median\", Norm) ~ score / medScore,\n",
    "            grepl(\"Max\", Norm) ~ score / maxScore,\n",
    "            grepl(\"MedMax\", Norm) ~ score / medMax,\n",
    "            TRUE ~ score / as.numeric(Norm)\n",
    "          )) %>%\n",
    "          plyranges::mutate(score = ifelse(grepl(\"MedMax\", Norm), score / medMax))\n",
    "      } else {\n",
    "        popCov <- tmp1 %>%\n",
    "          plyranges::mutate(Norm = rep(allNorm[[x]], n = plyranges::n())) %>%\n",
    "          plyranges::mutate(score = score / as.numeric(Norm))\n",
    "      }\n",
    "\n",
    "      exportCov(\n",
    "        CovObject = popCov, files = files,\n",
    "        fileName = gsub(\".bed\", \"_Insertions\", bedFiles2[x])\n",
    "      )\n",
    "    }, mc.cores = numCores)\n",
    "  }\n",
    "}\n",
    "\n",
    "\n",
    "#' Get basepair-resolution fragments coverage or insertion counts\n",
    "#'\n",
    "#' \\code{getbpCounts} Returns a GRanges or data frame containing single-basepair\n",
    "#' resolution scores on the given fragments.\n",
    "#'\n",
    "#' @param regionString The region from which to extract basepair counts. Format\n",
    "#'   should be as a string (e.g. 'chr1:1000-2000')\n",
    "#' @param popFrags Fragments from \\code{getPopFrags}\n",
    "#' @param numCores Number of cores to use.\n",
    "#' @param returnGRanges Set TRUE to return coverage as a GRanges, else default\n",
    "#'   behavior is to return a data frame\n",
    "#' @param sampleAvg Set to TRUE to take the sample average across a group. This\n",
    "#'   only works if \\code{popFrags} was generated with \\code{sampleSpecific=TRUE}\n",
    "#' @param approxLimit Max size of region in basepairs before binning is used.\n",
    "#' @param binSize Size of bins to use when binning\n",
    "#'\n",
    "#' @return A GRanges or Data Frame containing basepair resolution fragment or\n",
    "#'   insertion counts. If \\code{sampleAvg=TRUE} these counts will be averaged\n",
    "#'   over all samples within a population. If the region given by regionString\n",
    "#'   exceeds the size of approxLimit, counts will be aggregated into bins of\n",
    "#'   size binSize bp.\n",
    "getbpCounts <- function(regionString, popFrags, numCores,\n",
    "                        returnGRanges = FALSE, sampleAvg = TRUE,\n",
    "                        approxLimit = 100000, binSize = 250) {\n",
    "\n",
    "  # Parse through region\n",
    "  regionString <- gsub(\",\", \"\", regionString, fixed = TRUE)\n",
    "  chrom <- gsub(\":.*\", \"\", regionString)\n",
    "  startSite <- gsub(\".*:\", \"\", regionString) %>%\n",
    "    gsub(\"-.*\", \"\", .) %>%\n",
    "    as.numeric()\n",
    "  endSite <- gsub(\".*-\", \"\", regionString) %>% as.numeric()\n",
    "\n",
    "  regionGRanges <- GRanges(seqnames = chrom, ranges = IRanges(start = startSite, end = endSite), strand = \"*\")\n",
    "\n",
    "  # Extract cell counts for each group\n",
    "  groups <- gsub(\"__.*\", \"\", names(popFrags))\n",
    "  Norm <- gsub(\".*__\", \"\", names(popFrags))\n",
    "  names(Norm) <- groups\n",
    "\n",
    "  # Determine if binning is needed to simplify things\n",
    "  if (endSite - startSite > approxLimit) {\n",
    "    print(str_interp(\"Size of region exceeds ${approxLimit}bp. Binning data over ${binSize}bp windows.\"))\n",
    "    binnedData <- regionGRanges %>%\n",
    "      tile_ranges(., binSize) %>%\n",
    "      mutate(idx = c(1:length(.)))\n",
    "  }\n",
    "\n",
    "  # Summarize the coverage over the region window at a single basepair resolution\n",
    "  tmpCounts <- mclapply(seq_along(groups), function(x) {\n",
    "    print(paste(\"Counting\", names(popFrags)[x], sep = \" \"))\n",
    "\n",
    "    Num <- as.numeric(Norm[groups[x]])\n",
    "    tmpGR <- plyranges::compute_coverage(popFrags[[x]]) %>%\n",
    "      plyranges::mutate(score = score / Num) %>%\n",
    "      plyranges::filter_by_overlaps(regionGRanges)\n",
    "\n",
    "    # If we are binning, average scores over each bin\n",
    "    if (endSite - startSite > approxLimit) {\n",
    "      tmpGR <- plyranges::join_overlap_intersect(tmpGR, binnedData) %>%\n",
    "        plyranges::group_by(idx) %>%\n",
    "        plyranges::reduce_ranges(score = mean(score)) %>%\n",
    "        ungroup()\n",
    "    }\n",
    "\n",
    "    tmpGR\n",
    "  }, mc.cores = numCores)\n",
    "  names(tmpCounts) <- groups\n",
    "\n",
    "\n",
    "  if (all(lapply(seq_along(tmpCounts), function(x) {\n",
    "    sum(tmpCounts[[x]]$score) == 0\n",
    "  }))) {\n",
    "    stop(\"Error: No fragments detected in this region\")\n",
    "  }\n",
    "\n",
    "\n",
    "  # If you want to take the sample average across a group\n",
    "\n",
    "  if (sampleAvg == TRUE & all(grepl(\"#\", names(popFrags)))) {\n",
    "    groupCounts <- stack(as(lapply(tmpCounts, setSeqInfo), \"GRangesList\")) %>%\n",
    "      plyranges::mutate(Group = gsub(\"#.*\", \"\", name)) %>%\n",
    "      S4Vectors::split(., .$Group)\n",
    "\n",
    "    tmpCounts <- lapply(groupCounts, function(x) {\n",
    "      sampleCount <- length(unique(x$name))\n",
    "      x %>%\n",
    "        plyranges::compute_coverage(weight = .$score / sampleCount) %>%\n",
    "        plyranges::filter_by_overlaps(regionGRanges)\n",
    "    })\n",
    "    names(tmpCounts) <- names(groupCounts)\n",
    "  } else if (sampleAvg == TRUE) {\n",
    "    stop(\"Error: You asked for exporting sample averages for all cell groups, but didn't provide sample-specific fragment lists.\n",
    "\t\tPlease re-run getPopFrags with sampleSpecific = TRUE\")\n",
    "  }\n",
    "\n",
    "  if (returnGRanges) {\n",
    "    return(tmpCounts)\n",
    "  }\n",
    "\n",
    "  groups <- names(tmpCounts)\n",
    "\n",
    "\n",
    "  ## Generate a data.frame for export.\n",
    "  allTmp <- mclapply(seq_along(tmpCounts), function(x) {\n",
    "    tmpdf <- as.data.frame(tmpCounts[[x]])\n",
    "    if (dim(tmpdf)[1] < 2 & endSite - startSite <= approxLimit) {\n",
    "      data.frame(\n",
    "        chr = rep(chrom, (endSite - startSite + 1)),\n",
    "        Locus = c(startSite:endSite),\n",
    "        Counts = rep(0, (endSite - startSite + 1)),\n",
    "        CellType = rep(groups[x], (endSite - startSite + 1))\n",
    "      )\n",
    "\n",
    "      print(paste(groups[x], \" has no coverage for this region\", sep = \"\"))\n",
    "    } else if (dim(tmpdf)[1] < 2 & endSite - startSite > approxLimit) {\n",
    "\n",
    "      # If you bin data, then you need to set the Locus as the middle of the GRanges.\n",
    "      # This is only an issue if you aren't doing bp-specific coverage.\n",
    "      # In this case, if a given group has no coverage, then we create an empty\n",
    "      # data.frame for that group, using the midpoint of the whole region.\n",
    "      as_tibble(binnedData) %>%\n",
    "        dplyr::mutate(Groups = groups[x]) %>%\n",
    "        mutate(\n",
    "          Locus = as.integer((start + end) / 2),\n",
    "          Counts = 0,\n",
    "          chr = seqnames\n",
    "        ) %>%\n",
    "        dplyr::select(chr, Locus, Counts, Groups) %>%\n",
    "        dplyr::filter(Locus >= startSite & Locus <= endSite)\n",
    "\n",
    "      print(paste(groups[x], \" has no coverage for this region\", sep = \"\"))\n",
    "    } else if (endSite - startSite > approxLimit) {\n",
    "\n",
    "      # If you bin data, then you need to set the Locus as the middle of the GRanges.\n",
    "      # This is only an issue if you aren't doing bp-specific coverage.\n",
    "\n",
    "      as_tibble(tmpdf) %>%\n",
    "        dplyr::mutate(Groups = groups[x]) %>%\n",
    "        mutate(\n",
    "          Locus = as.integer((start + end) / 2),\n",
    "          Counts = score,\n",
    "          chr = seqnames\n",
    "        ) %>%\n",
    "        dplyr::select(chr, Locus, Counts, Groups) %>%\n",
    "        dplyr::filter(Locus >= startSite & Locus <= endSite)\n",
    "    } else {\n",
    "      tmp <- tmpCounts[[x]] %>% plyranges::tile_ranges(width = 1)\n",
    "      tmp$score <- tmpCounts[[x]]$score[tmp$partition]\n",
    "      tmp$Groups <- rep(groups[x], length(tmp))\n",
    "\n",
    "      covdf <- as.data.frame(tmp) %>% dplyr::select(seqnames, start, score, Groups)\n",
    "      colnames(covdf) <- c(\"chr\", \"Locus\", \"Counts\", \"Groups\")\n",
    "\n",
    "      covdf %>% dplyr::filter(Locus >= startSite & Locus <= endSite)\n",
    "    }\n",
    "  }, mc.cores = numCores)\n",
    "\n",
    "  tmpCountsdf <- as.data.frame(data.table::rbindlist(allTmp))\n",
    "\n",
    "  return(tmpCountsdf)\n",
    "}\n",
    "\n",
    "# same function as above but for insertions\n",
    "# This calculates bp counts over a region, scaled per 1000 cells\n",
    "# regionString = region that you want to plot, e.g. chr1:10000-20000\n",
    "# frags = list of fragments by group, derived from getPopFrags\n",
    "# numCores for multithreading\n",
    "\n",
    "getbpInserts <- function(regionString, popFrags, numCores = 1, sampleAvg = TRUE) {\n",
    "\n",
    "  # Parse through region\n",
    "  chrom <- gsub(\":.*\", \"\", regionString)\n",
    "  startSite <- gsub(\".*:\", \"\", regionString) %>%\n",
    "    gsub(\"-.*\", \"\", .) %>%\n",
    "    as.numeric()\n",
    "  endSite <- gsub(\".*-\", \"\", regionString) %>% as.numeric()\n",
    "\n",
    "  regionGRanges <- GRanges(seqnames = chrom, ranges = IRanges(start = startSite, end = endSite), strand = \"*\")\n",
    "\n",
    "  # Extract cell counts for each group\n",
    "  groups <- gsub(\"__.*\", \"\", names(popFrags))\n",
    "  Norm <- gsub(\".*__\", \"\", names(popFrags))\n",
    "  names(Norm) <- groups\n",
    "\n",
    "  tmpCounts <- lapply(c(1:length(popFrags)), function(x) {\n",
    "    print(paste(\"Counting insertions sites for \", names(popFrags)[x], sep = \" \"))\n",
    "    # Pull out insertion sites - start and end\n",
    "    cutstart <- GRanges(seqnames = as(seqnames(popFrags[[x]]), \"vector\"), ranges = IRanges(start = start(popFrags[[x]]), width = 1), strand = \"*\")\n",
    "    cutend <- GRanges(seqnames = as(seqnames(popFrags[[x]]), \"vector\"), ranges = IRanges(start = end(popFrags[[x]]), width = 1), strand = \"*\")\n",
    "    tmp_inserts <- plyranges::bind_ranges(cutstart, cutend)\n",
    "\n",
    "    # now iterate over region, looking at overlap with insertion sites\n",
    "    tmp <- unlist(mclapply(c(1:(endSite - startSite + 1)), function(i) {\n",
    "      singlebp <- GRanges(seqnames = chrom, ranges = IRanges(start = startSite + i, end = startSite + i), strand = \"*\")\n",
    "      bpRegion <- tmp_inserts %>% plyranges::join_overlap_intersect(singlebp)\n",
    "      length(bpRegion)\n",
    "    }, mc.cores = numCores))\n",
    "\n",
    "    Num <- as.numeric(Norm[groups[x]])\n",
    "    tmp / Norm\n",
    "  })\n",
    "  names(tmpCounts) <- groups\n",
    "  tmpdf <- as.data.frame(tmpCounts)\n",
    "  tmpdf$Locus <- c(startSite:endSite)\n",
    "\n",
    "  # Create a table of counts data by Celltype, and basepair, with a column for the Chromosome name.\n",
    "  tmpCountsdf <- tmpdf %>%\n",
    "    tidyr::pivot_longer(cols = colnames(.)[colnames(.) != \"Locus\"], names_to = \"Groups\", values_to = \"Counts\") %>%\n",
    "    dplyr::mutate(chr = rep(chrom, dim(.)[1]))\n",
    "\n",
    "\n",
    "  # If you want to take the sample average across a group\n",
    "\n",
    "  if (sampleAvg == TRUE & all(grepl(\"#\", names(popFrags)))) {\n",
    "    AvgCountsdf <- tmpCountsdf %>%\n",
    "      dplyr::mutate(Groups = gsub(\"#.*\", \"\", Groups)) %>%\n",
    "      group_by(Groups, chr, Locus) %>%\n",
    "      dplyr::summarize(Counts = mean(Counts))\n",
    "\n",
    "    return(AvgCountsdf)\n",
    "  } else if (sampleAvg == TRUE) {\n",
    "    stop(\"Error: You asked for exporting sample averages for all cell groups, but didn't provide sample-specific fragment lists.\n",
    "\t\tPlease re-run getPopFrags with sampleSpecific = TRUE\")\n",
    "  } else {\n",
    "    return(tmpCountsdf)\n",
    "  }\n",
    "}\n",
    "\n",
    "\n",
    "### Quick support functions for section 1 ---\n",
    "\n",
    "### Function to set seqinfo for GRanges objects.\n",
    "### This makes them easier to merge and export to BWs by setting the standard\n",
    "### chromosome lengths and names to a standard, as defined by TxDb.\n",
    "\n",
    "setSeqInfo <- function(GRangesObj, TxDb = TxDb.Hsapiens.UCSC.hg38.refGene) {\n",
    "  seqinfo(GRangesObj) <- seqinfo(TxDb)[seqnames(seqinfo(GRangesObj))]\n",
    "  return(GRangesObj)\n",
    "}\n",
    "\n",
    "\n",
    "## Export plyranges to various coveragefile formats\n",
    "\n",
    "exportCov <- function(CovObject, files, fileName, Txdb = TxDb.Hsapiens.UCSC.hg38.refGene) {\n",
    "\n",
    "  ### Let's exporting them to Bigwigs and/or bedgraphs\n",
    "\n",
    "  CovObject <- setSeqInfo(CovObject, Txdb)\n",
    "\n",
    "  print(paste(\"Exporting \", fileName, sep = \"\"))\n",
    "\n",
    "  if (all(tolower(files) == \"all\")) {\n",
    "\n",
    "    # Export BigWigs and BedGraphs\n",
    "\n",
    "    plyranges::write_bed_graph(CovObject, paste(fileName, \".bedGraph\", sep = \"\"))\n",
    "    plyranges::write_bigwig(CovObject, paste(fileName, \".bw\", sep = \"\"))\n",
    "  } else if (any(tolower(files) == \"bigwig\")) {\n",
    "\n",
    "    # Export Bigwigs only\n",
    "    plyranges::write_bigwig(CovObject, paste(fileName, \".bw\", sep = \"\"))\n",
    "  } else if (any(tolower(files) == \"bedgraph\")) {\n",
    "\n",
    "    # Export Bedgraphs only\n",
    "    plyranges::write_bed_graph(CovObject, paste(fileName, \".bedGraph\", sep = \"\"))\n",
    "  } else {\n",
    "    break(\"Error: Incorrect export option give. Check inputs\")\n",
    "  }\n",
    "}\n",
    "\n",
    "validRegionString <- function(regionString) {\n",
    "  if (!is.character(regionString)) {\n",
    "    return(FALSE)\n",
    "  }\n",
    "\n",
    "  pattern <- \"([0-9]{1,2}|chr[0-9]{1,2}):[0-9]*-[0-9]*\"\n",
    "  matchedPattern <- str_extract(regionString, pattern)\n",
    "\n",
    "  if (is.na(matchedPattern)) {\n",
    "    return(FALSE)\n",
    "  } else if (!matchedPattern == regionString) {\n",
    "    return(FALSE)\n",
    "  }\n",
    "\n",
    "  splits <- str_split(regionString, \"[:-]\")[[1]]\n",
    "  start <- splits[2]\n",
    "  end <- splits[3]\n",
    "  if (start > end) {\n",
    "    return(FALSE)\n",
    "  }\n",
    "\n",
    "  # All conditions satisfied\n",
    "  return(TRUE)\n",
    "}\n",
    "\n",
    "StringsToGRanges <- function(regionString) {\n",
    "  if (!validRegionString(regionString)) {\n",
    "    stop(\"Region must be a string matching format 'seqname:start-end', where start<end e.g. chr1:123000-123500\")\n",
    "  }\n",
    "\n",
    "  chrom <- gsub(\":.*\", \"\", regionString)\n",
    "  startSite <- gsub(\".*:\", \"\", regionString) %>%\n",
    "    gsub(\"-.*\", \"\", .) %>%\n",
    "    as.numeric()\n",
    "  endSite <- gsub(\".*-\", \"\", regionString) %>% as.numeric()\n",
    "\n",
    "\n",
    "  regionGRanges <- GRanges(seqnames = chrom, ranges = IRanges(start = startSite, end = endSite), strand = \"*\")\n",
    "  return(regionGRanges)\n",
    "}\n",
    "\n",
    "# Helper function to process and write to BAM\n",
    "writeBAMHelper <- function(popFrags, bedFiles, genome, numCores) {\n",
    "\n",
    "  # Coerce 'names' of popFrags to <character> for BamFile conversion\"\n",
    "  if (\"name\" %in% colnames(popFrags)) {\n",
    "    popFrags <- mclapply(seq_along(popFrags), function(x) {\n",
    "      if (\"name\" %in% colnames(popFrags)) {\n",
    "        popFrags[[x]] <- transform(popFrags[[x]], name = as.character(name))\n",
    "      }\n",
    "    }, mc.cores = numCores)\n",
    "  }\n",
    "\n",
    "  ## Fetch chromInfo file from UCSC for given genome, \"hg38\" default.\n",
    "  ## This uses fetchExtendedChromInfoFromUCSC() behind the scenes and\n",
    "  ## requires internet access. See ?fetchExtendedChromInfoFromUCSC\n",
    "  ## for the list of UCSC genomes that are currently supported.\n",
    "  chromInfo <- Seqinfo(genome = genome)\n",
    "\n",
    "  # Change bedFiles extensions to .bam\n",
    "  bamFiles <- gsub(\".bed\", \".bam\", bedFiles)\n",
    "\n",
    "  # Export data from GRanges to bam files, and end the function.\n",
    "  mclapply(seq_along(popFrags), function(x) {\n",
    "    # Export to bed\n",
    "    plyranges::write_bed(popFrags[[x]], bedFiles[x])\n",
    "\n",
    "    # Convert to GAlignments\n",
    "    popGAlignments <- as(popFrags[[x]], \"GAlignments\")\n",
    "\n",
    "    # Add seqinfo\n",
    "    chrNames <- names(seqlengths(popGAlignments))\n",
    "    seqinfo(popGAlignments) <- chromInfo[chrNames]\n",
    "\n",
    "    # Remove metadata cols\n",
    "    mcols(popGAlignments) <- NULL\n",
    "\n",
    "    # Export to BAM\n",
    "    rtracklayer::export(popGAlignments, bamFiles[x])\n",
    "  }, mc.cores = numCores)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70dcb192-d32a-41cf-9580-f5e2c59f1276",
   "metadata": {},
   "outputs": [],
   "source": [
    "######################### Section 2: Plotting Coverge and Insertion data for specific regions ----\n",
    "library(dplyr)\n",
    "library(tidyverse)\n",
    "library(plyranges)\n",
    "library(ggrepel)\n",
    "library(ggbio)\n",
    "library(TxDb.Hsapiens.UCSC.hg38.refGene)\n",
    "library(org.Hs.eg.db)\n",
    "library(OrganismDbi)\n",
    "library(doParallel)\n",
    "\n",
    "\n",
    "\n",
    "## Work Flow:\n",
    "#### 1. Pull out fragments for a given region by group - getPopFrags, from ArchR_Export_ATAC_Data.R\n",
    "######### source('ArchR_Export_ATAC_Data.R\")\n",
    "#### 2. Summarize the bp counts for either coverage or insertions, normalized per thousand cells\n",
    "#### 3. Plot coverage or insertions, with motif positions overlaid.\n",
    "\n",
    "\n",
    "# This calculates bp counts over a region, scaled per 1000 cells\n",
    "# region = region that you want to plot\n",
    "# regionString = region that you want to plot, e.g. chr1:10000-20000\n",
    "# numCores for multithreading\n",
    "\n",
    "\n",
    "##### supporting functions for plotRegion\n",
    "# Modified Track Plots\n",
    "\n",
    "#' Get gene promoter range string\n",
    "#'\n",
    "#' For a given gene, return the gene position from an ArchR project    \n",
    "#' \n",
    "#' @param proj An ArchR Project\n",
    "#' @param gene The name of a gene\n",
    "#' @param upstream The number of basepairs to extend the range upstream of the gene promoter\n",
    "#' @param downstream The number of basepairs to extend the range downstream of the gene promoter\n",
    "#' @param output_string Logical value, default TRUE. If TRUE outputs a range string (ie 'chr1: 1700000-1710000') else\n",
    "#' outputs a granges object\n",
    "#' \n",
    "#' @return A range string or granges object, depending on 'output_string' parameter.\n",
    "get_promoter_range <- function(proj, gene, upstream = 50000, downstream = 50000, output_string = TRUE) {\n",
    "  gene_pos <- ArchR::getGenes(proj, gene)\n",
    "  pro_pos_gr <- GenomicRanges::promoters(gene_pos, upstream = upstream, downstream = downstream)\n",
    "  if (output_string) {\n",
    "    pro_pos_str <- sprintf(\n",
    "      \"%s: %s-%s\",\n",
    "      GenomicRanges::seqnames(pro_pos_gr),\n",
    "      GenomicRanges::start(GenomicRanges::ranges(pro_pos_gr)),\n",
    "      GenomicRanges::end(GenomicRanges::ranges(pro_pos_gr))\n",
    "    )\n",
    "    return(pro_pos_str)\n",
    "  } else {\n",
    "    return(pro_pos_gr)\n",
    "  }\n",
    "}\n",
    "\n",
    "#' CountDF to Region GRanges\n",
    "#'\n",
    "#' Gets the range covered by counts in a count df and converts to granges\n",
    "#' Used in `plotRegion()`\n",
    "#'\n",
    "#' @param countdf  A dataframe that comes from `getbpCounts()` or `getbpInserts()`. Expected columns \"chr\" and \"Locus\"\n",
    "#' @return A granges object for the region defined in countdf\n",
    "countdf_to_region <- function(countdf) {\n",
    "  assertthat::assert_that(\"chr\" %in% names(countdf))\n",
    "  assertthat::assert_that(\"Locus\" %in% names(countdf))\n",
    "  \n",
    "  chrom <- toString(unique(countdf$chr))\n",
    "  startSite <- min(countdf$Locus)\n",
    "  endSite <- max(countdf$Locus)\n",
    "  regionGRanges <- GenomicRanges::GRanges(\n",
    "    seqnames = chrom,\n",
    "    ranges = IRanges::IRanges(start = startSite, end = endSite), strand = \"*\"\n",
    "  )\n",
    "  \n",
    "  return(regionGRanges)\n",
    "}\n",
    "\n",
    "#' Get GeneBody GRanges for a given region\n",
    "#'\n",
    "#' Used in `plotRegion()`\n",
    "#'\n",
    "#' @param regionGRanges regionGRanges A region Granges object to retrieve gene bodies for. For example, the output of countdf_to_region.\n",
    "#' @param TxDb A TxDb database, default TxDb.Hsapiens.UCSC.hg38.refGene\n",
    "#' @param single.strand.genes.only Logical, default FALSE.\n",
    "get_grange_genebody <- function(regionGRanges, TxDb = TxDb.Hsapiens.UCSC.hg38.refGene, single.strand.genes.only = FALSE) {\n",
    "  geneBody <- GenomicFeatures::genes(TxDb, single.strand.genes.only = single.strand.genes.only) %>%\n",
    "    unlist() %>%\n",
    "    plyranges::join_overlap_intersect(regionGRanges)\n",
    "  return(geneBody)\n",
    "}\n",
    "\n",
    "\n",
    "#' Default ggplot theme for counts plot\n",
    ".counts_plot_default_theme <- list(\n",
    "  panel.grid = element_blank(),\n",
    "  plot.margin = unit(c(0, 0, 0, 0), \"cm\"),\n",
    "  legend.title = element_text(size = 12),\n",
    "  legend.text = element_text(size = 10),\n",
    "  axis.text.y = element_blank(),\n",
    "  axis.ticks.y = element_blank(),\n",
    "  panel.border = element_blank(),\n",
    "  strip.background = element_blank(),\n",
    "  strip.text.x = element_text(size = 10, angle = 0),\n",
    "  strip.text.y = element_text(size = 10, angle = 0)\n",
    ")\n",
    "\n",
    "\n",
    "#' Plot normalized counts for each sample\n",
    "#'\n",
    "#' Used in `plotRegion()` for the counts tracks\n",
    "#'\n",
    "#' @param countdf  A dataframe that comes from `getbpCounts()` or `getbpInserts()`\n",
    "#' @param plotType Options include 'overlaid','stacked', or 'RidgePlot'. default is 'stacked', which will plot a seperate track for each group. \n",
    "#' Setting plotType to 'overlaid' will overlay count plot histograms across samples, instead of faceting out separately.\n",
    "#' Setting plotType to 'RidgePlot' will generate a ridgeplot across all groups. \n",
    "#' @param base_size Numeric, default 12. Global plot base text size parameter\n",
    "#' @param counts_color Optional color palette. A named vector of color values where names are unique values in the `color_var` column\n",
    "#' @param range_label_size Numeric value, default 4. Text size for the y-axis range label\n",
    "#' @param legend.position Any acceptable `legend.position` argument to theme(). Default NULL will place legend for overlaid plots at (0.8,0.8),\n",
    "#' or to the \"right\" for faceted plots.\n",
    "#' @param facet_label_side Direction character value, default \"top\". Can also be \"right\", \"left\", or \"bottom\". Position of facet label.\n",
    "#' @param counts_group_colors Optional named color vector. Values as colors, names are levels of `counts_color_var`. If provided, will color the plots specifically\n",
    "#' using `scale_color_manual()`\n",
    "#' @param counts_color_var Character value, default \"Groups\". Column name from countdf to use to color counts plots.\n",
    "#' Only used if counts_group_colors provided\n",
    "#' @param theme_ls A list of named theme arguments passed to theme(), defaults to `.counts_plot_default_theme`. For example, `list(axis.ticks = element_blank())`\n",
    "#' @return A ggplot object of count histograms by sample.\n",
    "\n",
    "counts_plot_samples <- function(countdf,\n",
    "                                plotType = 'stacked',\n",
    "                                base_size = 12,\n",
    "                                counts_color = NULL,\n",
    "                                range_label_size = 4,\n",
    "                                legend.position = NULL,\n",
    "                                facet_label_side = \"top\",\n",
    "                                counts_group_colors = NULL,\n",
    "                                counts_color_var = \"Groups\",\n",
    "                                theme_ls = .counts_plot_default_theme) {\n",
    "  assertthat::assert_that(\"Counts\" %in% names(countdf))\n",
    "  assertthat::assert_that(\"Locus\" %in% names(countdf))\n",
    "  assertthat::assert_that(\"Groups\" %in% names(countdf))\n",
    "  assertthat::assert_that(counts_color_var %in% names(countdf))\n",
    "  \n",
    "  # Fill in theme any unspecified theme options with defaults\n",
    "  default_theme <- .counts_plot_default_theme\n",
    "  unspec_param <- setdiff(names(default_theme), names(theme_ls))\n",
    "  if (length(unspec_param) > 0) {\n",
    "    theme_ls <- c(theme_ls, default_theme[unspec_param])\n",
    "  }\n",
    "  \n",
    "  # Range df for plotting\n",
    "  df_range <- data.frame(\n",
    "    x = max(countdf$Locus),\n",
    "    y = max(countdf$Counts),\n",
    "    label = paste0(\"Range:\", 0, \"-\", round(max(countdf$Counts), digits = 2))\n",
    "  )\n",
    "  \n",
    "  # Intialize plot\n",
    "  p1 <- ggplot(data = countdf, aes(x = Locus, y = Counts)) +\n",
    "    theme_bw(base_size = base_size)\n",
    "  \n",
    "  \n",
    "  # Plots\n",
    "  if (tolower(plotType) == 'overlaid') {\n",
    "    # Conditional Theme\n",
    "    if (is.null(legend.position)) {\n",
    "      legend.position <- c(0.8, 0.8)\n",
    "    }\n",
    "    theme_ls <- c(\n",
    "      theme_ls,\n",
    "      list(\n",
    "        legend.position = legend.position\n",
    "      )\n",
    "    )\n",
    "    \n",
    "    # Base Plot\n",
    "    p1 <- p1 +\n",
    "      geom_line(aes(color = !!as.name(counts_color_var)), alpha = 0.75, size = 1.5) +\n",
    "      ylab(NULL) +\n",
    "      labs(Groups = \"Groups\") +\n",
    "      coord_cartesian(clip = \"off\") +\n",
    "      geom_text(\n",
    "        data = df_range,\n",
    "        aes(x = x, y = y, label = label),\n",
    "        size = range_label_size,\n",
    "        hjust = 1,\n",
    "        vjust = 1\n",
    "      ) +\n",
    "      do.call(theme, theme_ls)\n",
    "    \n",
    "    # Conditional Plot Elements\n",
    "    if (!is.null(counts_group_colors)) {\n",
    "      # assertthat::assert_that(all(unique(countdf[[counts_color_var]]) %in% names(counts_group_colors)),\n",
    "      #                        msg = \"Must supply colors for all levels of color variable\")\n",
    "      p1 <- p1 + scale_color_manual(values = counts_group_colors, breaks = names(counts_group_colors))\n",
    "    }\n",
    "  } else if(tolower(plotType) == 'stacked'){\n",
    "    # Conditional Theme\n",
    "    if (is.null(legend.position)) {\n",
    "      legend.position <- \"right\"\n",
    "    }\n",
    "    theme_ls <- c(\n",
    "      theme_ls,\n",
    "      list(\n",
    "        legend.position = legend.position\n",
    "      )\n",
    "    )\n",
    "    \n",
    "    # Base Plot, conditional\n",
    "    if (is.null(counts_color)) {\n",
    "      p1 <- p1 +\n",
    "        geom_area(aes(fill = !!as.name(counts_color_var)), position = \"identity\")\n",
    "    } else {\n",
    "      p1 <- p1 +\n",
    "        geom_area(fill = counts_color, position = \"identity\")\n",
    "    }\n",
    "    \n",
    "    # Base Plot, common elements\n",
    "    p1 <- p1 +\n",
    "      ylab(NULL) +\n",
    "      facet_wrap(vars(Groups), ncol = 1, strip.position = facet_label_side) +\n",
    "      geom_text(\n",
    "        data = df_range,\n",
    "        aes(x = x, y = y, label = label),\n",
    "        size = range_label_size,\n",
    "        hjust = 1,\n",
    "        vjust = 1\n",
    "      ) +\n",
    "      do.call(theme, theme_ls)\n",
    "    \n",
    "    # Conditional Plot Elements\n",
    "    if (!is.null(counts_group_colors)) {\n",
    "      # commenting this section out so we can supply specific colors to highlight just a subset if we want--other samples will just be gray\n",
    "      # assertthat::assert_that(all(unique(countdf[[counts_color_var]]) %in% names(counts_group_colors)),\n",
    "      #                        msg = \"Must supply colors for all levels of color variable\")\n",
    "      p1 <- p1 +\n",
    "        scale_fill_manual(values = counts_group_colors, breaks = names(counts_group_colors))\n",
    "    }\n",
    "  } else if(tolower(plotType) == 'ridgeplot'){\n",
    "    \n",
    "    \n",
    "    ##Conditional elements needed to make RidgePlots work.\n",
    "    countdf_tmp<- as.data.frame(table(countdf$Groups)) \n",
    "    countdf$Groups2 <- match(countdf$Groups,countdf_tmp$Var1)\n",
    "    \n",
    "    \n",
    "  #Base plot, conditional\n",
    "   p1 <- p1 + \n",
    "      geom_ridgeline(data = as.data.frame(countdf),\n",
    "                     aes(x = Locus, y = Groups2, height = Counts, \n",
    "                         fill = Groups),\n",
    "                     alpha = 0.25)+\n",
    "      ylab(NULL) + \n",
    "      scale_y_continuous(breaks=c(1:length(tmp$Var1)), \n",
    "                        label= tmp$Var1) +\n",
    "      geom_text(\n",
    "        data = data.frame(\n",
    "          x = Inf, y = Inf,\n",
    "          label = paste(\"Range:\", 0, \"-\", round(max(countdf$Counts), digits = 2), sep = \"\")\n",
    "        ),\n",
    "        aes(x = x, y = y, label = label), \n",
    "        size = 2, hjust = 0.9, vjust = 1.4\n",
    "      ) + \n",
    "      theme(legend.position = \"none\")\n",
    "    \n",
    "    \n",
    "  } else {stop(\"Error: Plot type not recognized. Please check input for variable 'plotType'\")}\n",
    "  \n",
    "  return(p1)\n",
    "}\n",
    "\n",
    "#' Get scaled values for custom breaks\n",
    "#'\n",
    "#' Get scaled values for custom breaks in a given value vector for use\n",
    "#' defining breaks in scale_gradientn(), for example\n",
    "#' @param breaks\n",
    "#' @param x\n",
    "breaks_to_scaledbreaks <- function(breaks, x) {\n",
    "  rescaled_weights <- scales::rescale(x)\n",
    "  rescaled_breaks <- quantile(rescaled_weights, probs = ecdf(x)(breaks))\n",
    "  return(rescaled_breaks)\n",
    "}\n",
    "\n",
    "cleanup_breaks <- function(breaks, x) {\n",
    "  breaks <- sort(breaks)\n",
    "  n_lower <- sum(breaks < min(x))\n",
    "  while (n_lower > 1) {\n",
    "    breaks <- breaks[-1]\n",
    "    n_lower <- sum(breaks < min(x))\n",
    "  }\n",
    "  n_higher <- sum(breaks > max(x))\n",
    "  while (n_higher > 1) {\n",
    "    breaks <- breaks[-length(breaks)]\n",
    "    n_higher <- sum(breaks > max(x))\n",
    "  }\n",
    "  return(breaks)\n",
    "}\n",
    "\n",
    "\n",
    "#' Get GRanges of motif annotations within a region\n",
    "#'\n",
    "#' Helper to return the motif annotations within a specfic region outside of\n",
    "#' the plotting function. Can supply either a counts df or a region string.\n",
    "#'\n",
    "#' @param ArchRProj The archr project containing motif annotations\n",
    "#' @param motifSetName The name of the motif annotations in the ArchR project\n",
    "#' @param countdf A counts data frame object from getPop\n",
    "get_motifs_in_region <- function(\n",
    "  ArchRProj, \n",
    "  motifSetName, \n",
    "  countdf = NULL, \n",
    "  regionString = NULL, \n",
    "  numCores = 1, \n",
    "  metaColumn = NULL, \n",
    "  cellSubsets = \"ALL\"\n",
    ") {\n",
    "  \n",
    "  if (is.null(countdf) & !is.null(regionString)) {\n",
    "    popFrags <- getPopFrags(ArchRProj = ArchRProj, metaColumn = metaColumn, region = regionString, numCores = numCores, cellSubsets = cellSubsets)\n",
    "    countdf <- getbpCounts(regionString = regionString, popFrags = popFrags, numCores = numCores, returnGRanges = FALSE)\n",
    "  }\n",
    "  chrom <- toString(unique(countdf$chr))\n",
    "  startSite <- min(countdf$Locus)\n",
    "  endSite <- max(countdf$Locus)\n",
    "  regionGRanges <- GRanges(seqnames = chrom, ranges = IRanges(start = startSite, end = endSite), strand = \"*\")\n",
    "  \n",
    "  specMotifs <- unlist(getPositions(ArchRProj, name = motifSetName)) %>%\n",
    "    plyranges::mutate(name = gsub(\"_.*\", \"\", names(.))) %>%\n",
    "    plyranges::join_overlap_intersect(regionGRanges) %>%\n",
    "    plyranges::mutate(index = seq(1, length(.), by = 1)) %>%\n",
    "    plyranges::mutate(labels = paste(name, index, sep = \"_\"))\n",
    "  \n",
    "  return(specMotifs)\n",
    "}\n",
    "\n",
    "#' Plot Motifs on Counts Plot\n",
    "#'\n",
    "#' Add motif annotation labels on an existing counts plot. If providing weights\n",
    "#' for motif text colors, ensure the input counts plot is formatted accordingly\n",
    "#' (ie counts_color = \"gray90\") for best visibility.\n",
    "#'\n",
    "#' @param p1 The output of `counts_plot_samples()`\n",
    "#' @param countdf  A dataframe that comes from `getbpCounts()` or\n",
    "#'   `getbpInserts()`\n",
    "#' @param ArchRProj An archR project containing motif annotations\n",
    "#' @param motifSetName The name of the motif set in ArchRProj to use for\n",
    "#'   annotation. Example: 'JasparMotifs'\n",
    "#' @param customFeatures Optional - A GRanges containing a user-specific set of\n",
    "#'   motifs or other features. If specified these will be used instead of the\n",
    "#'   ArchR motif set.\n",
    "#' @param motif_y_space_factor A factor for vertical spacing between motif\n",
    "#'   labels. Default 4. Increase to make labels farther apart, decrease to make\n",
    "#'   labels closer.\n",
    "#' @param motif_stagger_labels_y = FALSE Logical value, default FALSE. If TRUE,\n",
    "#'   will  stagger motif labels in adjacent columns in the vertical direction\n",
    "#' @param motif_weights Optional numeric vector, default NULL. If provided will\n",
    "#'   be used to color motif labels by the weighted values. Values must be\n",
    "#'   uniquely named with motif names, for example c(`KLF5`= 3.2, `STAT1 = 0.2`,\n",
    "#'   `EOMES` = -1.4`). Weights can be anything relevant, for example if the\n",
    "#'   peak/region is associated with a specific group/sample then global motif\n",
    "#'   enrichment results for that group: `-log10(FDR)*sign(change)`\n",
    "#' @param motif_weight_name Character value, default \"Motif Weight\". Used to\n",
    "#'   label the color legend.\n",
    "#' @param weight_colors Named numeric vector. Names should be color values and\n",
    "#'   breaks should be the corresponding values of motif_weights. Values outside\n",
    "#'   the highest and lowest value will appear as max or min defined color value.\n",
    "#' @param motif_lab_size Numeric value, default 1. Size of motif labels.\n",
    "#' @param motif_lab_alpha Numeric value, default 0.25. Alpha for motif labels.\n",
    "#' @param motif_line_size Numeric value, default 1. Size of motif lines.\n",
    "#' @param motif_line_alpha Numeric value, default 0.25. Alpha for motif lines.\n",
    "#' @return The input ggplot object with motif labels overlaid\n",
    "counts_plot_motif_overlay <- function(p1,\n",
    "                                      countdf,\n",
    "                                      ArchRProj = NULL,\n",
    "                                      motifSetName = NULL,\n",
    "                                      customFeatures = NULL,\n",
    "                                      motif_y_space_factor = 4,\n",
    "                                      motif_stagger_labels_y = FALSE,\n",
    "                                      motif_weights = NULL,\n",
    "                                      motif_weight_name = \"Motif Weight\",\n",
    "                                      motif_weight_colors = c(darkblue = -10, gray = 0, darkred = 10),\n",
    "                                      motif_lab_size = 1,\n",
    "                                      motif_lab_alpha = 0.25,\n",
    "                                      motif_line_size = 0.75,\n",
    "                                      motif_line_alpha = 0.25) {\n",
    "  \n",
    "  # Retrieve annotations in region and format\n",
    "  if (!is.null(customFeatures)){\n",
    "    \n",
    "    # Use custom features/motifs\n",
    "    specMotifs <- customFeatures\n",
    "    # Filter these to our region\n",
    "    chrom <- toString(unique(countdf$chr))\n",
    "    startSite <- min(countdf$Locus)\n",
    "    endSite <- max(countdf$Locus)\n",
    "    regionGRanges <- GRanges(seqnames = chrom, ranges = IRanges(start = startSite, end = endSite), strand = \"*\")\n",
    "    specMotifs <- specMotifs %>%\n",
    "      plyranges::mutate(name = gsub(\"_.*\", \"\", names(.))) %>%\n",
    "      plyranges::join_overlap_intersect(regionGRanges) %>%\n",
    "      plyranges::mutate(index = seq(1, length(.), by = 1)) %>%\n",
    "      plyranges::mutate(labels = paste(name, index, sep = \"_\"))\n",
    "    \n",
    "  } else if (!is.null(motifSetName) & !is.null(ArchRProj)){\n",
    "    \n",
    "    # Get our motifs from the ArchR Project\n",
    "    assertthat::assert_that(\n",
    "      motifSetName %in% names(ArchRProj@peakAnnotation),\n",
    "      msg = sprintf(\"%s not found in ArchRProj\", motifSetName)\n",
    "    )\n",
    "    specMotifs <- get_motifs_in_region(\n",
    "      countdf = countdf,\n",
    "      ArchRProj = ArchRProj,\n",
    "      motifSetName = motifSetName\n",
    "    )\n",
    "    \n",
    "  } else {\n",
    "    stop(\"To plot motifs either motifSetName and ArchRProj must be provided, or provide a GRanges to customFeatures\")\n",
    "  }\n",
    "  \n",
    "  reduceMotifs <- plyranges::reduce_ranges(specMotifs, count = plyranges::n(), names = paste(labels, collapse = \",\"))\n",
    "  splitMotifs <- strsplit(reduceMotifs$names, split = \",\")\n",
    "  \n",
    "  # Calculate y spacing for motif labels\n",
    "  minSize <- max(countdf$Counts) / (length(specMotifs) / motif_y_space_factor)\n",
    "  y1 <- rep(0, length(specMotifs))\n",
    "  names(y1) <- specMotifs$labels\n",
    "  \n",
    "  for (i in 1:length(splitMotifs)) {\n",
    "    if (i %% 2 == 0 & motif_stagger_labels_y) { # stagger y spacing in neighboring columns\n",
    "      start_val <- minSize / 2\n",
    "    } else {\n",
    "      start_val <- 0\n",
    "    }\n",
    "    \n",
    "    y1[splitMotifs[[i]]] <- seq(start_val, start_val + minSize * (length(splitMotifs[[i]]) - 1), by = minSize)\n",
    "  }\n",
    "  \n",
    "  # TF label coordinates and labels\n",
    "  tmp_motifdf <- data.frame(\n",
    "    x1 = start(specMotifs),\n",
    "    x2 = start(specMotifs) + width(specMotifs),\n",
    "    y = y1,\n",
    "    name = specMotifs$labels\n",
    "  ) %>%\n",
    "    pivot_longer(cols = c(\"x1\", \"x2\"), names_to = NULL, values_to = \"x\") %>%\n",
    "    group_by(name) %>%\n",
    "    mutate(labels = ifelse(max(x) == x, gsub(\"_.*\", \"\", name), NA))\n",
    "  \n",
    "  # Incorporate Weights\n",
    "  if (!is.null(motif_weights)) {\n",
    "    assertthat::assert_that(is.numeric(motif_weight_colors))\n",
    "    \n",
    "    if (length(intersect(names(motif_weights), tmp_motifdf$labels)) == 0) {\n",
    "      warning(sprintf(\n",
    "        \"None of the supplied motif weight names match expected motif labels. Example motif label format: %s\",\n",
    "        paste(head(unique(na.omit(tmp_motifdf$label)), 3), collapse = \", \")\n",
    "      ))\n",
    "    }\n",
    "    tmp_motifdf <- tmp_motifdf %>%\n",
    "      mutate(mweight = ifelse(labels %in% names(motif_weights), motif_weights[labels], NA)) %>%\n",
    "      mutate(mweight = ifelse(!all(is.na(mweight)), max(mweight, na.rm = T), NA)) %>% # still grouped by name. this ensures weights are applied to each row of same motif\n",
    "      mutate(mweight = case_when( # clip ends within color range\n",
    "        mweight > max(motif_weight_colors) ~ max(motif_weight_colors),\n",
    "        mweight < min(motif_weight_colors) ~ min(motif_weight_colors),\n",
    "        is.na(mweight) ~ as.numeric(NA),\n",
    "        TRUE ~ mweight\n",
    "      ))\n",
    "  } else {\n",
    "    tmp_motifdf$mweight <- 1\n",
    "  }\n",
    "  \n",
    "  # Plot\n",
    "  p1 <- p1 +\n",
    "    geom_line(\n",
    "      data = tmp_motifdf,\n",
    "      aes(x = x, y = y, group = name, color = mweight),\n",
    "      alpha = motif_line_alpha,\n",
    "      size = motif_line_size\n",
    "    ) +\n",
    "    # vertical labels\n",
    "    ggrepel::geom_text_repel(\n",
    "      data = tmp_motifdf[tmp_motifdf$y > 0 & !is.na(tmp_motifdf$labels), ], # removed the NA rows to prevent warnings for intentionally missing labels\n",
    "      aes(x = x, y = y, label = labels, color = mweight),\n",
    "      direction = \"x\",\n",
    "      arrow = arrow(length = unit(0.5, \"mm\")),\n",
    "      alpha = motif_lab_alpha,\n",
    "      size = motif_lab_size,\n",
    "      segment.size = 0.25,\n",
    "      max.overlaps = 50,\n",
    "      min.segment.length = 0\n",
    "    ) +\n",
    "    # motif labels along x-axis\n",
    "    ggrepel::geom_text_repel(\n",
    "      data = tmp_motifdf[tmp_motifdf$y == 0 & !is.na(tmp_motifdf$labels), ], # removed the NA rows to prevent warnings for intentionally missing labels\n",
    "      aes(x = x, y = y, label = labels, color = mweight),\n",
    "      arrow = arrow(length = unit(0.5, \"mm\")),\n",
    "      alpha = motif_lab_alpha,\n",
    "      size = motif_lab_size,\n",
    "      direction = \"y\",\n",
    "      segment.size = 0.25,\n",
    "      min.segment.length = 0,\n",
    "      nudge_y = -max(countdf$Counts) / 20\n",
    "    ) +\n",
    "    ylim(-max(countdf$Counts) / 10, max(countdf$Counts)) +\n",
    "    xlim(min(countdf$Locus), max(countdf$Locus))\n",
    "  \n",
    "  # Color text by motif weights\n",
    "  if (!is.null(motif_weights)) {\n",
    "    scaled_breaks <- breaks_to_scaledbreaks(motif_weight_colors, na.omit(tmp_motifdf$mweight))\n",
    "    \n",
    "    p1 <- p1 + scale_color_gradientn(\n",
    "      colors = names(motif_weight_colors), values = scaled_breaks,\n",
    "      na.value = \"gray\", # may want to parameterize this\n",
    "      name = motif_weight_name\n",
    "    )\n",
    "  } else {\n",
    "    p1 <- p1 + scale_color_gradient(low = \"black\", high = \"black\", na.value = \"gray\", name = motif_weight_name, guide = \"none\")\n",
    "  }\n",
    "  \n",
    "  return(p1)\n",
    "}\n",
    "\n",
    "\n",
    "#' Plot colors in a color palette\n",
    "#'\n",
    "#' Utility function to quickly visualize a color palette formatted for motif weight visualization\n",
    "#'\n",
    "#' @param pal A palette, named list of values where names are colors and values are the color breaks\n",
    "#' @return Plots a base plot to visualize the input palette\n",
    "#' @examples\n",
    "#' motif_color_pal <- c(-20, -10, 0, 10, 20)\n",
    "#' names(motif_color_pal) <- c(\"darkblue\", \"cyan\", \"gray\", \"orange\", \"darkred\")\n",
    "#' plot_pal(motif_color_pal)\n",
    "plot_pal <- function(pal, cex = 4, cex.lab = 1) {\n",
    "  npal <- length(pal)\n",
    "  plotdim(npal * 1, 3)\n",
    "  plot(1:npal, rep(1, npal), pch = 19, cex = cex, col = names(pal))\n",
    "  text(1:npal, rep(1, npal), labels = pal, cex = cex.lab)\n",
    "}\n",
    "\n",
    "#' Get Gene Body Model\n",
    "#'\n",
    "#' Get Gene Body model for specific gene in region ranges\n",
    "#'\n",
    "#' Used in `plotRegion()` to preformat GRanges for selected genes\n",
    "#'\n",
    "#' @param whichGene Character value. Gene symbol for a single gene to plot.\n",
    "#' @param regionGRanges Granges object for region of interest. Ie, an output of `countdf_to_region()`\n",
    "#' @param countdf  A dataframe that comes from `getbpCounts()` or `getbpInserts()`\n",
    "#' @param orgdb An organism database containing the gene, default `org.Hs.eg.db`\n",
    "#' @param db_id_col Character value. Column in `orgdb` containing the output id. Default \"REFSEQ\".\n",
    "#' @param verbose Logical value, default TRUE.\n",
    "#' @return GRanges object for gene to plot\n",
    "get_gene_body_model <- function(whichGene,\n",
    "                                regionGRanges,\n",
    "                                countdf,\n",
    "                                orgdb = org.Hs.eg.db,\n",
    "                                db_id_col = \"REFSEQ\",\n",
    "                                verbose = TRUE) {\n",
    "  # Get REFSEQ values for gene symbol\n",
    "  # this requires package \"RMariaDB\" available on CRAN and corresponding linux library \"libmariadb-dev\"\n",
    "  txList <- tryCatch(\n",
    "    unlist(AnnotationDbi::mapIds(\n",
    "      x = orgdb,\n",
    "      keys = whichGene,\n",
    "      column = db_id_col,\n",
    "      keytype = \"SYMBOL\",\n",
    "      multiVals = \"list\"\n",
    "    )),\n",
    "    error = function(e) {\n",
    "      stop(sprintf(\n",
    "        \"Error attempting to query symbol for 'whichGene' argument [%s] from specified database. Full error message: %s\",\n",
    "        paste(whichGene, collapse = \",\"), e\n",
    "      ))\n",
    "      return(NULL)\n",
    "    }\n",
    "  )\n",
    "  \n",
    "  # Pull specified database and table for the transcripts by\n",
    "  newTxDB <- GenomicFeatures::makeTxDbFromUCSC(genome = \"hg38\", tablename = \"refGene\", transcript_ids = txList)\n",
    "  if (verbose) {\n",
    "    print(txList)\n",
    "  }\n",
    "  \n",
    "  # Get granges object from database and intersect with region granges\n",
    "  newTxDBgenes <- unlist(genes(newTxDB, single.strand.genes.only = FALSE)) %>%\n",
    "    plyranges::filter(!grepl(\"_\", seqnames))\n",
    "  newGRanges <- biovizBase::crunch(newTxDB, which = newTxDBgenes) %>%\n",
    "    plyranges::join_overlap_intersect(regionGRanges)\n",
    "  colnames(values(newGRanges))[4] <- \"model\" # rename 'Type' column to models\n",
    "  \n",
    "  if (length(newGRanges) > 0) {\n",
    "    newGRangesf <- as.data.frame(newGRanges)\n",
    "    startSide <- newGRangesf[newGRangesf$start %in% start(regionGRanges) & newGRangesf$model == \"gap\", ]\n",
    "    endSide <- newGRangesf[newGRangesf$end %in% end(regionGRanges) & newGRangesf$model == \"gap\", ]\n",
    "    beginexon <- NULL\n",
    "    endexon <- NULL\n",
    "    \n",
    "    if (dim(startSide)[1] > 0) {\n",
    "      beginexon <- data.frame(\n",
    "        seqnames = unique(countdf$chr),\n",
    "        start = min(countdf$Locus) - 1,\n",
    "        end = min(countdf$Locus) - 1,\n",
    "        width = 1,\n",
    "        strand = \"*\",\n",
    "        tx_id = unique(startSide$tx_id),\n",
    "        tx_name = unique(startSide$tx_name),\n",
    "        gene_id = unique(startSide$gene_id),\n",
    "        model = \"exon\"\n",
    "      )\n",
    "    }\n",
    "    \n",
    "    if (dim(endSide)[1] > 0) {\n",
    "      endexon <- data.frame(\n",
    "        seqnames = unique(countdf$chr),\n",
    "        start = max(countdf$Locus) + 1,\n",
    "        end = max(countdf$Locus) + 1,\n",
    "        width = 1,\n",
    "        strand = \"*\",\n",
    "        tx_id = unique(endSide$tx_id),\n",
    "        tx_name = unique(endSide$tx_name),\n",
    "        gene_id = unique(endSide$gene_id),\n",
    "        model = \"exon\"\n",
    "      )\n",
    "    }\n",
    "    \n",
    "    newModel <- rbind(newGRangesf, beginexon, endexon) %>%\n",
    "      mutate(nameList = paste(whichGene, tx_name, sep = \": \")) %>%\n",
    "      makeGRangesListFromDataFrame(., split.field = \"nameList\", keep.extra.columns = TRUE)\n",
    "    \n",
    "    # check for transcripts that are identical within the window\n",
    "    uniqueTranscripts <- stack(newModel)[!duplicated(ranges(stack(newModel)))] %>%\n",
    "      plyranges::filter(model != \"utr\") %>%\n",
    "      as.data.frame() %>%\n",
    "      dplyr::select(tx_name) %>%\n",
    "      unique(.)\n",
    "    newModel <- newModel[grepl(paste(unlist(uniqueTranscripts), collapse = \"|\"), names(newModel))]\n",
    "    \n",
    "    names(newModel) <- rep(whichGene, length(newModel))\n",
    "    \n",
    "    return(newModel)\n",
    "  } else {\n",
    "    print(\"Warning: whichGene did not contain genes names for any genes in the window. Plotting all transcripts\")\n",
    "    return(NULL)\n",
    "  }\n",
    "}\n",
    "\n",
    "#' Common theme for gene plots\n",
    ".gene_plot_theme <- list(\n",
    "  panel.grid = element_blank(),\n",
    "  panel.border = element_blank(),\n",
    "  axis.ticks = element_blank(),\n",
    "  axis.text.y = element_blank(),\n",
    "  plot.margin = unit(c(.5, .5, .5, .5), \"cm\")\n",
    ")\n",
    "\n",
    "#' Generate a ggplot based on a Gene Model\n",
    "#'\n",
    "#' Creates a gene track ggplot based on gene model GRanges\n",
    "#' Used conditinally in `plotRegion()` for gene plot.\n",
    "#'\n",
    "#' @param A gene granges model, ie generated by `get_gene_body_model()`\n",
    "#' @param x_min Numeric value. Min x-axis limit. If NULL (default) will use min value from newModel granges.\n",
    "#' @param x_max Numeric value. Max x-axis limit. If NULL (default) will use max value from newModel granges.\n",
    "#' @param base_size Numeric, default 12. Global plot base text size parameter\n",
    "#' @param theme_ls Named list of parameters passed to `theme()`. For defaults see `.gene_plot_theme`\n",
    "plot_whichGene <- function(newModel,\n",
    "                           x_lim = NULL,\n",
    "                           base_size = 12,\n",
    "                           theme_ls = .gene_plot_theme) {\n",
    "  # if(is.null(x_min)){\n",
    "  #     xmin = min(newModel$\n",
    "  # }\n",
    "  \n",
    "  # Fill in theme any unspecified theme options with defaults\n",
    "  default_theme <- .gene_plot_theme\n",
    "  unspec_param <- setdiff(names(default_theme), names(theme_ls))\n",
    "  if (length(unspec_param) > 0) {\n",
    "    theme_ls <- c(theme_ls, default_theme[unspec_param])\n",
    "  }\n",
    "  \n",
    "  p_gene <- ggbio() +\n",
    "    geom_alignment(newModel,\n",
    "                   aes(type = model),\n",
    "                   cds.rect.h = 0.25,\n",
    "                   rect.height = 0.25 / 4\n",
    "    )\n",
    "  \n",
    "  if (!is.null(x_lim)) {\n",
    "    p_gene <- p_gene + xlim(x_lim[1], x_lim[2])\n",
    "  }\n",
    "  \n",
    "  p_gene <- p_gene +\n",
    "    scale_y_continuous(expand = c(0.3, 0.3)) +\n",
    "    theme_bw(base_size = base_size) +\n",
    "    coord_cartesian(clip = \"off\") +\n",
    "    do.call(theme, theme_ls)\n",
    "  \n",
    "  return(p_gene)\n",
    "}\n",
    "\n",
    "#' Plot a genebody from a specific database\n",
    "#'\n",
    "#' Used in `plotRegion()` to plot gene track for granges input containing all genes in a region\n",
    "#'\n",
    "#' @param organismdb Database object, for example the output of `OrganismDbi::makeOrganismDbFromTxDb()`\n",
    "#' @param geneBody_gr A GRanges object containing the gene, for example an output of `get_grange_genebody()`\n",
    "#' @param xmin\n",
    "#' @param xmax\n",
    "#' @param base_size Numeric, default 12. Global plot base text size parameter\n",
    "#' @param plot_theme Named list of `ggplot2::theme()` parameters.\n",
    "#' @param collapseGenes Logical value, default FALSE. If TRUE will collapse all genes into one line.\n",
    "#' sites into a single row\n",
    "#' @return A ggplot object\n",
    "plot_geneBody <- function(organismdb,\n",
    "                          geneBody_gr,\n",
    "                          x_lim = NULL,\n",
    "                          base_size = 12,\n",
    "                          theme_ls = .gene_plot_theme,\n",
    "                          collapseGenes = FALSE) {\n",
    "  # Fill in theme any unspecified theme options with defaults\n",
    "  default_theme <- .gene_plot_theme\n",
    "  unspec_param <- setdiff(names(default_theme), names(theme_ls))\n",
    "  if (length(unspec_param) > 0) {\n",
    "    theme_ls <- c(theme_ls, default_theme[unspec_param])\n",
    "  }\n",
    "  \n",
    "  if (collapseGenes) {\n",
    "    stat_val <- \"reduce\"\n",
    "  } else {\n",
    "    stat_val <- NULL\n",
    "  }\n",
    "  \n",
    "  g_geneBody <- ggbio::autoplot(organismdb, wh = geneBody_gr)#, stat = stat_val)\n",
    "  \n",
    "  if (!is.null(x_lim)) { # had to move this here, otherwise theme would be overwritten...\n",
    "    g_geneBody <- g_geneBody + xlim(x_lim[1], x_lim[2])\n",
    "  }\n",
    "  \n",
    "  g_geneBody <- g_geneBody +\n",
    "    scale_y_continuous(expand = c(0.3, 0.3)) +\n",
    "    theme_bw(base_size = base_size) +\n",
    "    coord_cartesian(clip = \"off\") +\n",
    "    do.call(theme, theme_ls)\n",
    "  \n",
    "  \n",
    "  return(g_geneBody)\n",
    "}\n",
    "\n",
    "##### simplifiedOrgDb: Generates a simplified OrganismDb object for plotting only the longest transcript for each gene. \n",
    "\n",
    "#' Generate a simplified Organism Db object.\n",
    "#'\n",
    "#' Pulls out only the longest transcripts for each gene and repackages them into an organism database object for plotting. \n",
    "#'\n",
    "#' @param TxDb A TxDb object with transcript info. \n",
    "#' @param orgdb A OrgDb object with gene name info. \n",
    "\n",
    "simplifiedOrgDb <- function(TxDb = TxDb, orgdb = orgdb){\n",
    "  \n",
    "  # retrieve transcript lengths\n",
    "  txlen <- transcriptLengths(TxDb, with.utr5_len=TRUE, with.utr3_len=TRUE)\n",
    "  setDT(txlen)\n",
    "  txlen$len <- rowSums(as.matrix(txlen[, .(tx_len, utr5_len, utr3_len)]))\n",
    "  \n",
    "  setkey(txlen, gene_id, len, tx_id)\n",
    "  \n",
    "  # filter longesttranscript by gene_id\n",
    "  ltx <- txlen[!is.na(gene_id)][, tail(.SD,1), by=gene_id]$tx_id\n",
    "  \n",
    "  # filter txdb object\n",
    "  txb <- as.list(TxDb)\n",
    "  txb$transcripts <- txb$transcripts[txb$transcripts$tx_id %in% ltx, ]\n",
    "  txb$splicings <- txb$splicings[txb$splicings$tx_id %in% ltx,]\n",
    "  txb$genes <- txb$genes[txb$genes$tx_id %in% ltx,]\n",
    "  chrominfo <- as.data.frame(seqinfo(TxDb)) %>%\n",
    "    dplyr::mutate(chrom = rownames(.)) %>% \n",
    "    dplyr::rename(length = seqlengths, is_circular = isCircular)\n",
    "  txb2 <- GenomicFeatures::makeTxDb(txb$transcripts, txb$splicings, txb$genes, chrominfo, metadata = \n",
    "                                      filter(metadata(TxDb), \n",
    "                                             grepl('Genome|Organism|Data source|UCSC Table',name)))\n",
    "  Homo.sapiens.hg38 <- OrganismDbi::makeOrganismDbFromTxDb(txb2, \n",
    "                                                           orgdb = orgdb)\n",
    "  \n",
    "  return(Homo.sapiens.hg38)\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "##### plotRegion: Plots the region that you've summarized across all cell groupings.\n",
    "\n",
    "# Preformatting for Motif Overlay:\n",
    "# Use any of these lines of code to add a motif position set to an ArchR Project with peak calls\n",
    "### proj<- addMotifAnnotations(ArchRProj = proj, motifSet = \"JASPAR2020\", name = \"JasparMotifs\")\n",
    "########### Other options: motifSet = \"cisbp\", \"encode\", \"homor\".\n",
    "########### name = \"whatever name you want here\". This is \"name\" field is what you need to pass to plotRegion to access the motif positions.\n",
    "\n",
    "#' Plot Region\n",
    "#'\n",
    "#' Plots the region that you've summarized across all cell groupings\n",
    "#' (groups=initial getPopFrags() split) with optional motif overlay, chromosome\n",
    "#' position ideogram, and additional GRanges tracks. If plotting motif overlay,\n",
    "#' ensure that motif annotations have been added to your project peak calls.\n",
    "#' Example: `proj<- addMotifAnnotations(ArchRProj = proj, motifSet =\n",
    "#' \"JASPAR2020\", name = \"JasparMotifs\")`. A basic plot can be rendered with just\n",
    "#' a counts data frame, but additional formatting arguments allow for further\n",
    "#' customization.\n",
    "#'\n",
    "#' @param countdf A dataframe that comes from `getbpCounts()` or\n",
    "#'   `getbpInserts()`\n",
    "#' @param plotType Options include 'overlaid','stacked', or 'RidgePlot'. default\n",
    "#'   is 'stacked', which will plot a seperate track for each group. Setting\n",
    "#'   plotType to 'overlaid' will overlay count plot histograms across samples,\n",
    "#'   instead of faceting out separately. Setting plotType to 'RidgePlot' will\n",
    "#'   generate a ridgeplot across all groups.\n",
    "#' @param base_size Numeric, default 12. Global plot base text size parameter\n",
    "#' @param counts_color Optional color palette. A named vector of color values\n",
    "#'   where names are unique values in the `color_var` column\n",
    "#' @param range_label_size Numeric value, default 4. Text size for the y-axis\n",
    "#'   range label\n",
    "#' @param legend.position Any acceptable `legend.position` argument to theme().\n",
    "#'   Default NULL will place legend for overlaid plots at (0.8,0.8), or to the\n",
    "#'   \"right\" for faceted plots.\n",
    "#' @param facet_label_side Direction character value, default \"top\". Can also be\n",
    "#'   \"right\", \"left\", or \"bottom\". Position of facet label.\n",
    "#' @param counts_group_colors Optional named color vector. Values as colors,\n",
    "#'   names are levels of `counts_color_var`. If provided, will color the plots\n",
    "#'   specifically using `scale_color_manual()`\n",
    "#' @param counts_color_var Character value, default \"Groups\". Column name from\n",
    "#'   countdf to use to color counts plots. Only used if counts_group_colors\n",
    "#'   provided\n",
    "#' @param counts_theme_ls A list of named theme arguments passed to theme(). For\n",
    "#'   example, `list(axis.ticks = element_blank())`. Default NULL will use\n",
    "#'   `.counts_plot_default_theme`.\n",
    "#' @param ArchRProj An archR project containing motif annotations. Only needed\n",
    "#'   if supplying `motifSetName` arg for motif label overlay.\n",
    "#' @param motifSetName The name of the motif set in ArchRProj to use for\n",
    "#'   annotation. Example: 'JasparMotifs'\n",
    "#' @param customFeatures Optional - A GRanges containing a user-specific set of\n",
    "#'   motifs or other features. If specified these will be used instead of the\n",
    "#'   ArchR motif set.\n",
    "#' @param motif_y_space_factor A factor for vertical spacing between motif\n",
    "#'   labels. Default 4. Increase to make labels farther apart, decrease to make\n",
    "#'   labels closer.\n",
    "#' @param motif_stagger_labels_y = FALSE Logical value, default FALSE. If TRUE,\n",
    "#'   will  stagger motif labels in adjacent columns in the vertical direction\n",
    "#' @param motif_weights Optional numeric vector, default NULL. If provided will\n",
    "#'   be used to color motif labels by the weighted values\n",
    "#' @param motif_weight_name Character value, default \"Motif Weight\". Used to\n",
    "#'   label the legend for motif colors\n",
    "#' @param weight_colors Named numeric vector. Names should be color values and\n",
    "#'   breaks should be the corresponding values of motif_weights. Values outside\n",
    "#'   the highest and lowest value will appear as max or min defined color value.\n",
    "#' @param motif_lab_size Numeric value, default 1. Size of motif labels.\n",
    "#' @param motif_lab_alpha Numeric value, default 0.25. Alpha for motif labels.\n",
    "#' @param motif_line_size Numeric value, default 1. Size of motif lines.\n",
    "#' @param motif_line_alpha Numeric value, default 0.25. Alpha for motif lines.\n",
    "#' @param showGene Logical value, default TRUE. Whether or not the gene track\n",
    "#'   should be plotted.\n",
    "#' @param TxDb A TxDb (transcript database) object used for gene body references\n",
    "#'   when autoplotting all genes in region. Default\n",
    "#'   `TxDb.Hsapiens.UCSC.hg38.refGene`, UCSC Ref Gene from Hg38.\n",
    "#' @param whichGene Optional character value. Gene symbol for a single gene to\n",
    "#'   plot instead of all genes in region. Default NULL.\n",
    "#' @param orgdb An organism database containing the gene regions for `whichGene`\n",
    "#'   plotting, default `org.Hs.eg.db`\n",
    "#' @param db_id_col Character value. Column in `orgdb` containing the output id\n",
    "#'   for `whichGene` plotting. Default \"REFSEQ\".\n",
    "#' @param collapseGenes Options include 'collapseAll', 'longestTx', or 'None'\n",
    "#'   Default 'None' will plot the expanded view of the reference genes,\n",
    "#'   'collapseAll' if you want collapse the gene tracks into one, and\n",
    "#'   'longestTx' will only plot the longest transcript of each gene.\n",
    "#' @param gene_theme_ls Named list of parameters passed to `theme()` for the\n",
    "#'   gene plot. Default NULL will use `.gene_plot_theme`\n",
    "#' @param additionalGRangesTrack A GRanges object containing additional track\n",
    "#'   plot data\n",
    "#' @param showIdeogram Logical value, default TRUE. If TRUE plots the chromosome\n",
    "#'   ideogram at the top of the multi-track plot\n",
    "#' @param ideogram_genome Character value, a genome name for the ideogram plot.\n",
    "#'   Default 'hg19'.\n",
    "#' @param relativeHeights Named numeric vector of relative heights for each of\n",
    "#'   the 4 track plots to enable clean visualization when there are many tracks.\n",
    "#'   Unused tracks will be ignored. Default value = c(`Chr` = 0.9, `Normalized\n",
    "#'   Counts` = 7, `Genes`= 2, `AdditionalGRanges` = 4.5)\n",
    "#' @param verbose Logical value, default FALSE. If TRUE will show all messages\n",
    "#'   generated by internal function calls. If FALSE, will attempt to use\n",
    "#'   suppressMessages() to minimize message output.\n",
    "#' @return The input ggplot object with motif labels overlaid\n",
    "#' @examples\n",
    "#' \\dontrun{\n",
    "#' # my_count_df is a counts data frame generated by getbpCounts()\n",
    "#'\n",
    "#' # Simple counts + ideogram + all genes:\n",
    "#' plotRegion(countdf = my_count_df)\n",
    "#'\n",
    "#' # Motif overlay for a project (my_proj) containing \"JasparMotifs\" annotations:\n",
    "#' plotRegion(countdf = my_count_df, ArchRProj = my_proj, motifSetName = \"JasparMotifs\", motif_lab_alpha = 1, motif_line_alpha = 1)\n",
    "#' }\n",
    "#'\n",
    "#' # Motif overlay w/ weights.:\n",
    "#' plotRegion(\n",
    "#'   countdf = my_count_df, ArchRProj = my_proj, motifSetName = \"JasparMotifs\", motif_lab_alpha = 1,\n",
    "#'   motif_line_alpha = 1, motif_weights = my_enrichment_weights\n",
    "#' )\n",
    "#'\n",
    "plotRegion <- function(countdf,\n",
    "                       # base count plot args\n",
    "                       plotType = 'stacked',\n",
    "                       base_size = 12,\n",
    "                       counts_color = NULL,\n",
    "                       range_label_size = 2,\n",
    "                       legend.position = NULL,\n",
    "                       facet_label_side = \"top\",\n",
    "                       counts_color_var = \"Groups\",\n",
    "                       counts_group_colors = NULL,\n",
    "                       counts_theme_ls = NULL,\n",
    "                       # Motif args\n",
    "                       ArchRProj = NULL, # For overlaying motifs only\n",
    "                       motifSetName = NULL,\n",
    "                       customFeatures = NULL,\n",
    "                       motif_y_space_factor = 4,\n",
    "                       motif_stagger_labels_y = FALSE,\n",
    "                       motif_weights = NULL,\n",
    "                       motif_weight_name = \"Motif Weight\",\n",
    "                       motif_weight_colors = c(darkblue = -10, gray = 0, darkred = 10),\n",
    "                       motif_lab_size = 1,\n",
    "                       motif_lab_alpha = 0.25,\n",
    "                       motif_line_alpha = 0.25,\n",
    "                       motif_line_size = 0.75,\n",
    "                       # Genes plot args\n",
    "                       showGene = TRUE,\n",
    "                       TxDb = TxDb.Hsapiens.UCSC.hg38.refGene, #\n",
    "                       whichGene = NULL, # for plotting specific gene in region\n",
    "                       orgdb = org.Hs.eg.db, \n",
    "                       db_id_col = \"REFSEQ\",\n",
    "                       collapseGenes = 'None',\n",
    "                       gene_theme_ls = NULL,\n",
    "                       single.strand.genes.only = FALSE,\n",
    "                       # Additional Tracks\n",
    "                       additionalGRangesTrack = NULL,\n",
    "                       loopdf = NULL,\n",
    "                       # Ideogram\n",
    "                       showIdeogram = TRUE,\n",
    "                       ideogram_genome = \"hg19\",\n",
    "                       # Combined Tracks\n",
    "                       relativeHeights = c(`Chr` = 0.9, `Normalized Counts` = 7, `Genes` = 2, `AdditionalGRanges` = 4.5),\n",
    "                       verbose = FALSE) {\n",
    "  # Validate input\n",
    "  supported_tracks <- c(\"Chr\", \"Normalized Counts\", \"Genes\", \"AdditionalGRanges\")\n",
    "  if (length(setdiff(names(relativeHeights), supported_tracks)) > 0) {\n",
    "    warning(sprintf(\n",
    "      \"1 or more values of relative heights not in supported tracks: %s.\\n Supported track names: %s\",\n",
    "      paste(setdiff(names(relativeHeights), supported_tracks), collapse = \", \"),\n",
    "      paste(supported_tracks, collapse = \", \")\n",
    "    ))\n",
    "  }\n",
    "  \n",
    "  # function wrapper based on verbosity to hide messages\n",
    "  verbf <- function(x) {\n",
    "    if (verbose) {\n",
    "      x\n",
    "    } else {\n",
    "      suppressMessages(x)\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  # Variables\n",
    "  chrom <- toString(unique(countdf$chr))\n",
    "  .relativeHeights_default <- c(`Chr` = 0.9, `Normalized Counts` = 7, `Genes` = 2, `AdditionalGRanges` = 4.5,  `Loops` = 1.5) # retain in case any missing\n",
    "  \n",
    "  # Extract region from countdf as granges\n",
    "  regionGRanges <- countdf_to_region(countdf = countdf)\n",
    "  \n",
    "  # Base Plot of Sample Counts\n",
    "  p1 <- verbf(\n",
    "    counts_plot_samples(countdf,\n",
    "                        plotType = plotType,\n",
    "                        base_size = base_size,\n",
    "                        counts_color_var = counts_color_var,\n",
    "                        counts_color = counts_color,\n",
    "                        range_label_size = range_label_size,\n",
    "                        legend.position = legend.position,\n",
    "                        facet_label_side = facet_label_side,\n",
    "                        counts_group_colors = counts_group_colors,\n",
    "                        theme_ls = counts_theme_ls\n",
    "    )\n",
    "  )\n",
    "  \n",
    "  # Add Motifs to Base Plot if Requested\n",
    "  if (\n",
    "    (!is.null(motifSetName) & !is.null(ArchRProj)) || !is.null(customFeatures)\n",
    "  ) {\n",
    "    p1 <- verbf(\n",
    "      counts_plot_motif_overlay(\n",
    "        p1 = p1,\n",
    "        countdf = countdf,\n",
    "        ArchRProj = ArchRProj,\n",
    "        motifSetName = motifSetName,\n",
    "        customFeatures = customFeatures,\n",
    "        motif_y_space_factor = motif_y_space_factor,\n",
    "        motif_stagger_labels_y = motif_stagger_labels_y,\n",
    "        motif_weights = motif_weights,\n",
    "        motif_weight_name = motif_weight_name,\n",
    "        motif_weight_colors = motif_weight_colors,\n",
    "        motif_lab_size = motif_lab_size,\n",
    "        motif_lab_alpha = motif_lab_alpha,\n",
    "        motif_line_alpha = motif_line_alpha,\n",
    "        motif_line_size = motif_line_size\n",
    "      )\n",
    "    )\n",
    "  } else if (!is.null(motifSetName) & is.null(ArchRProj)) {\n",
    "    stop(\"No ArchR Project found. If you wish to plot motif positions overlaid, please provide an ArchR project along with your motifSetName.\")\n",
    "  } else {\n",
    "    p1 <- verbf(\n",
    "      p1 + xlim(min(countdf$Locus), max(countdf$Locus))\n",
    "    )\n",
    "  }\n",
    "  \n",
    "  # Build P2, Ref Genes track\n",
    "  if (showGene) {\n",
    "    # If user provided a specific gene symbol(s), pull new granges from database and format models\n",
    "    if (!is.null(whichGene)) {\n",
    "      newModel <- verbf(\n",
    "        get_gene_body_model(\n",
    "          whichGene = whichGene,\n",
    "          countdf = countdf,\n",
    "          regionGRanges = regionGRanges,\n",
    "          orgdb = orgdb,\n",
    "          verbose = verbose\n",
    "        )\n",
    "      )\n",
    "    } else {\n",
    "      newModel <- NULL\n",
    "    }\n",
    "    \n",
    "    \n",
    "    if (!is.null(whichGene) & !is.null(newModel)) {\n",
    "      # Successful newmodel generated\n",
    "      p2 <- verbf(\n",
    "        plot_whichGene(newModel,\n",
    "                       base_size = base_size,\n",
    "                       x_lim = range(countdf$Locus),\n",
    "                       theme_ls = gene_theme_ls\n",
    "        )\n",
    "      )\n",
    "    } else {\n",
    "      # Create your reference for plotting gene body\n",
    "      \n",
    "      if(tolower(collapseGenes) == 'longesttx'){\n",
    "        \n",
    "        Homo.sapiens.hg38 <- simplifiedOrgDb(TxDb = TxDb, orgdb = orgdb)\n",
    "        \n",
    "      }else{\n",
    "        \n",
    "        Homo.sapiens.hg38 <- verbf(OrganismDbi::makeOrganismDbFromTxDb(TxDb, orgdb = orgdb))\n",
    "      }\n",
    "      \n",
    "      geneBody <- verbf(get_grange_genebody(regionGRanges, TxDb = TxDb, single.strand.genes.only = single.strand.genes.only))\n",
    "   \n",
    "      if (length(geneBody) > 0) {\n",
    "        p2 <- verbf(\n",
    "          plot_geneBody(\n",
    "            organismdb = Homo.sapiens.hg38,\n",
    "            geneBody_gr = geneBody,\n",
    "            collapseGenes = tolower(collapseGenes) == 'collapseall',\n",
    "            base_size = base_size,\n",
    "            x_lim = range(countdf$Locus),\n",
    "            theme_ls = gene_theme_ls\n",
    "          )\n",
    "        )\n",
    "      } else {\n",
    "        print(\"No gene body in range\")\n",
    "        showGene <- FALSE # to prevent errors resulting from p2 nonexistence\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## TODO process additional granges\n",
    "  if (!is.null(additionalGRangesTrack)) {\n",
    "    \n",
    "    # Check for name metadata column\n",
    "    if (\"name\" %in% colnames(mcols(additionalGRangesTrack))) {\n",
    "      # Only plot the overlap of this region and the additional GRanges Track\n",
    "      overlapGRanges <- verbf(plyranges::join_overlap_intersect(additionalGRangesTrack, regionGRanges))\n",
    "      if (length(overlapGRanges) > 0) {\n",
    "        # Use the subset within our region as the track we want to plot\n",
    "        print(\"GRanges has overlap\")\n",
    "        # Assign exon status to each row\n",
    "        mcols(overlapGRanges)$type <- rep(\"exon\", each = length(overlapGRanges))\n",
    "        # split into list of GRanges -> GRangesList named for the names metadata col\n",
    "        additionalGRangesTrack <- split(overlapGRanges, overlapGRanges$name)\n",
    "      } else {\n",
    "        print(length(overlapGRanges))\n",
    "        print(\"No overlap of additional GRanges and this region\")\n",
    "        additionalGRangesTrack <- NULL\n",
    "        # Avoids the following error:\n",
    "        # Error: Faceting variables must have at least one value\n",
    "        # Error in deparse(x[[i]], nlines = nlines) :   no slot of name \"elementType\" for this object of class \"GRanges\"\n",
    "      }\n",
    "    } else {\n",
    "      print(\"Additional GRanges does not have `name` in metadata - ignoring additional track\")\n",
    "      additionalGRangesTrack <- NULL\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  ## Generate Loop track\n",
    "  if(!is.null(loopdf)){\n",
    "    \n",
    "    p5 <- ggplot() + \n",
    "            geom_curve(aes(x = midPeak1, xend = midPeak2,\n",
    "                                    y = y, yend = y, color = Correlation), \n",
    "                                curvature = -.5,\n",
    "                                data = cbind(loopdf,y= rep(0, dim(loopdf)[1]))) +\n",
    "           theme_minimal() + ylab(NULL) + xlab(NULL) +  ylim(-1, 0) +\n",
    "           scale_colour_viridis_c(breaks = c(ceiling(10*min(loopdf$Correlation))/10,\n",
    "                                        0,\n",
    "                                        floor(10*max(loopdf$Correlation))/10))  +\n",
    "            coord_cartesian(ylim = c(-0.25, 0))  +\n",
    "      theme(panel.grid = element_blank(), panel.border = element_blank(),\n",
    "             axis.text.x = element_blank(),  axis.ticks.x = element_blank(),\n",
    "             axis.text.y = element_blank(),  axis.ticks.y = element_blank())\n",
    "    \n",
    "  }\n",
    "  \n",
    "  # Combine plots P1...P4\n",
    "  # Base tracks are:\n",
    "  # 1: Chromosome Ideogram\n",
    "  # 2: a) Normalized Counts\n",
    "  # 2: b) Optional: Loop track\n",
    "  # 3: Ref Genes\n",
    "  # Additional GRanges track is user-defined and labeled according to its\n",
    "  # 'names' metadata column\n",
    "  # 4: AdditionalGRanges\n",
    "  ## TODO\n",
    "  \n",
    "  # Construct Named plot list top to Bottom by appending to track list if present\n",
    "  track_list <- list()\n",
    "  \n",
    "  if (showIdeogram) {\n",
    "    # Show Ideogram\n",
    "    p3 <- verbf(Ideogram(genome = ideogram_genome, subchr = chrom))\n",
    "    \n",
    "    track_list <- c(track_list, list(\"Chr\" = p3))\n",
    "  }\n",
    "  \n",
    "  # Counts\n",
    "  track_list <- c(track_list, list(\"Normalized Counts\" = p1))\n",
    "  \n",
    "  # Loops\n",
    "  if (!is.null(loopdf)) {\n",
    "    track_list <- c(track_list, list(\"Loops\" = p5))\n",
    "  }\n",
    "  \n",
    "  # Genes\n",
    "  if (showGene) {\n",
    "    track_list <- c(track_list, list(\"Genes\" = p2))\n",
    "  }\n",
    "  \n",
    "  # Additional Ranges\n",
    "  if (!is.null(additionalGRangesTrack)) {\n",
    "    print(\"Combining base tracks with an additional GRanges track\")\n",
    "    p4 <- verbf(autoplot(additionalGRangesTrack))\n",
    "    track_list <- c(track_list, list(\"AdditionalGRanges\" = p4))\n",
    "  }\n",
    "  \n",
    "  # height params\n",
    "  if (!all(names(track_list) %in% names(relativeHeights))) {\n",
    "    missing_heights <- setdiff(names(track_list), names(relativeHeights))\n",
    "    append_heights <- .relativeHeights_default[missing_heights]\n",
    "    relativeHeights <- c(relativeHeights, append_heights)\n",
    "    warning(sprintf(\n",
    "      \"Relative heights were not defined for included plots [%s]. Using defaults for these tracks [%s]\",\n",
    "      paste(missing_heights, collapse = \", \"),\n",
    "      paste(append_heights, collapse = \", \")\n",
    "    ))\n",
    "  }\n",
    "  trackHeights <- relativeHeights[names(track_list)] # ensure intended order\n",
    "  # Plot All Supplied Plots\n",
    "  g_tracks <- verbf(\n",
    "    ggbio::tracks(\n",
    "      track_list,\n",
    "      heights = trackHeights,\n",
    "      xlim = range(countdf$Locus),\n",
    "      track.bg.color = \"transparent\",\n",
    "      padding = unit(-1, \"lines\"),\n",
    "      label.bg.fill = \"transparent\",\n",
    "      label.bg.color = \"transparent\",\n",
    "      label.width = unit(2, \"lines\")\n",
    "    )\n",
    "    # coord_cartesian(clip = \"off\")\n",
    "  )\n",
    "  \n",
    "  return(g_tracks)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d13fce03-9f69-451b-9a1b-5ba063e724dd",
   "metadata": {},
   "source": [
    "## IFNG Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd0415f3-be42-4d5c-9b8e-1ceb5ef344b8",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "pop_frags <- getPopFrags(ArchRProj = proj_subset, metaColumn = 'pure_clusters', sampleSpecific = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c105ec5-d9a7-4967-a541-b58df3b84c0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "ifng <- getbpCounts(regionString = \"chr12:68,153,524-68,160,983\", popFrags = pop_frags, numCores = 16)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a4bb91f-82d9-40ff-8f45-26c84bf3bca2",
   "metadata": {},
   "source": [
    "## Plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "845fa020-de33-4899-9603-3ff47c94d690",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 5, repr.plot.height = 10)\n",
    "plotRegion(countdf = ifng, ArchRProj = proj_subset, counts_group_colors = c('#980043','#006837','#3690c0'))\n",
    "# '#006837','#3690c0','#980043'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f221dab2-c1b8-4830-acf5-ad8fb767cb5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Open a pdf file\n",
    "pdf(\"plots/cd4na_celltypes_ifng_track.pdf\", width = 5, height = 10) \n",
    "# 2. Create a plot\n",
    "plotRegion(countdf = ifng, ArchRProj = proj_subset, counts_group_colors = c('#980043','#006837','#3690c0'))\n",
    "# Close the pdf file\n",
    "dev.off() "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dc79fbe3-b09b-43b7-9b04-a4876c573968",
   "metadata": {},
   "source": [
    "# CD4 Naive Tracks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68fbb504-6fd3-4a9f-b8fa-a6fc9259e17c",
   "metadata": {},
   "outputs": [],
   "source": [
    "idxNaive <- BiocGenerics::which(proj_subset$pure_clusters %in% \"CD4 Naive\")\n",
    "cellsNaive <- proj_subset$cellNames[idxNaive]\n",
    "proj_naive <- proj_subset[cellsNaive, ]\n",
    "proj_naive"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c20dd2e-9bc6-4c8b-9583-4794f2a3a384",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "pop_frags_naive <- getPopFrags(ArchRProj = proj_naive, metaColumn = 'age_celltype', sampleSpecific = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e087cd6f-591f-422a-8a38-5b963bb59cbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "sox4 <- getbpCounts(regionString = \"chr6:21,592,533-21,599,836\", popFrags = pop_frags_naive, numCores = 16)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "abeb622f-208c-491c-aeda-f9c2299bc343",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 10, repr.plot.height = 6)\n",
    "plotRegion(countdf = sox4, ArchRProj = proj_naive, counts_group_colors = c('#1b9e77','#d95f02'))\n",
    "# counts_group_colors = c('#f46d43','#d73027','#2ca25f')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01d91cbf-ef0c-4fa0-b283-f7d15bc2c042",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Open a pdf file\n",
    "pdf(\"plots/cd4na_celltypes_sox4_track.pdf\", width = 10, height = 6) \n",
    "# 2. Create a plot\n",
    "plotRegion(countdf = sox4, ArchRProj = proj_naive, counts_group_colors = c('#1b9e77','#d95f02'))\n",
    "# Close the pdf file\n",
    "dev.off() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d96b402-34b5-444d-b506-6639f68e81b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "cpq <- getbpCounts(regionString = \"chr8:96,642,827-96,647,013\", popFrags = pop_frags_naive, numCores = 16)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4391659-9bd3-4738-a22b-c86fa17aef7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 10, repr.plot.height = 6)\n",
    "plotRegion(countdf = cpq, ArchRProj = proj_naive, counts_group_colors = c('#1b9e77','#d95f02'))\n",
    "# counts_group_colors = c('#f46d43','#d73027','#2ca25f')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0088b110-7a52-4d0c-87eb-b6c9f82525c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Open a pdf file\n",
    "pdf(\"plots/cd4na_celltypes_cpq_track.pdf\", width = 10, height = 6) \n",
    "# 2. Create a plot\n",
    "plotRegion(countdf = cpq, ArchRProj = proj_naive, counts_group_colors = c('#1b9e77','#d95f02'))\n",
    "# Close the pdf file\n",
    "dev.off() "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46028a92-cc69-4224-b6aa-379b427f45d6",
   "metadata": {},
   "source": [
    "# Pediatric vs Adult CD4NA Volcano"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c56a759-74dc-4f46-ad1d-1def214cd11e",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "markerPeaks_MNP <- getMarkerFeatures(proj_subset, groupBy = \"age_celltype\", useGroups = \"Pediatric_CD4 Naive\", bgdGroups = \"Senior_CD4 Naive\", maxCells = 10000,\n",
    "useMatrix = \"PeakMatrix\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1236e6f9-8fa1-4b26-9777-480bc4022d8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "cisbpEnrich1 <- peakAnnoEnrichment(\n",
    "seMarker = markerPeaks_MNP,\n",
    "ArchRProj = proj_subset,\n",
    "peakAnnotation = \"Motif\",\n",
    "cutOff = \"FDR <= 0.1 & Log2FC < 0\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1aa63e93-209d-49ed-8cee-529a747abd44",
   "metadata": {},
   "outputs": [],
   "source": [
    "cisbprEnrich2 <- peakAnnoEnrichment(\n",
    "seMarker = markerPeaks_MNP,\n",
    "ArchRProj = proj_subset,\n",
    "peakAnnotation = \"Motif\",\n",
    "cutOff = \"FDR <= 0.1 & Log2FC >= 0\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3dffd06-3d1d-4ba9-9625-98d3961ef957",
   "metadata": {},
   "outputs": [],
   "source": [
    "JDF1 <- as.data.frame(assays(cisbpEnrich1)) %>% mutate(TF = gsub(\"_.*\", \"\", rownames(.))) %>%\n",
    "tidyr::pivot_wider(id_cols = TF, names_from = group_name, values_from=value) %>%\n",
    "mutate(Enrichment = -as.numeric(Enrichment), mlog10Padj = as.numeric(mlog10Padj), Group = \"Adult CD4Na\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60170226-73ed-4d98-baa4-9810859762ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "JDF2 <- as.data.frame(assays(cisbprEnrich2)) %>% mutate(TF = gsub(\"_.*\", \"\", rownames(.))) %>%\n",
    "tidyr::pivot_wider(id_cols = TF, names_from = group_name, values_from=value) %>%\n",
    "mutate(Enrichment = as.numeric(Enrichment), mlog10Padj = as.numeric(mlog10Padj), Group = \"Pediatric CD4Na\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f2bdc36-436e-4cc0-8389-9782c383e865",
   "metadata": {},
   "outputs": [],
   "source": [
    "JDF <- rbind(JDF1, JDF2) %>% mutate(delabel = ifelse(abs(Enrichment) > 1 & mlog10Padj > 15, TF, NA))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1424e92-52a3-40a4-8d7d-976838c502e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "head(JDF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4011924d-2658-4de2-8b00-a73d88259931",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(ggrepel)\n",
    "library(scales)\n",
    "squish_trans <- function(from, to, factor) {\n",
    "  \n",
    "  trans <- function(x) {\n",
    "    \n",
    "    if (any(is.na(x))) return(x)\n",
    "\n",
    "    # get indices for the relevant regions\n",
    "    isq <- x > from & x < to\n",
    "    ito <- x >= to\n",
    "    \n",
    "    # apply transformation\n",
    "    x[isq] <- from + (x[isq] - from)/factor\n",
    "    x[ito] <- from + (to - from)/factor + (x[ito] - to)\n",
    "    \n",
    "    return(x)\n",
    "  }\n",
    "\n",
    "  inv <- function(x) {\n",
    "    \n",
    "    if (any(is.na(x))) return(x)\n",
    "\n",
    "    # get indices for the relevant regions\n",
    "    isq <- x > from & x < from + (to - from)/factor\n",
    "    ito <- x >= from + (to - from)/factor\n",
    "    \n",
    "    # apply transformation\n",
    "    x[isq] <- from + (x[isq] - from) * factor\n",
    "    x[ito] <- to + (x[ito] - (from + (to - from)/factor))\n",
    "    \n",
    "    return(x)\n",
    "  }\n",
    "  \n",
    "  # return the transformation\n",
    "  return(trans_new(\"squished\", trans, inv))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c07d22d9-f0fd-4121-beb5-aee08230beb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(repr.plot.width = 14, repr.plot.height = 8)\n",
    "ggplot(JDF, aes(x = Enrichment, y= mlog10Padj, label=delabel)) + geom_point(aes(color = Group)) + theme_minimal() +\n",
    "geom_vline(xintercept=c(-1, 1), col=\"red\") +\n",
    "scale_color_manual(values = c('#d95f02','#1b9e77')) +\n",
    "geom_hline(yintercept=5, col=\"red\") + ggtitle(\"CISBP Motif Enrichment of CD4Na Age\") +\n",
    "coord_cartesian(clip = \"off\") + scale_x_continuous(trans = squish_trans(-1,1,20),\n",
    "                                                    breaks = c(-1.5,-1.25,-1,0,1,1.25,1.5),\n",
    "                                                    labels = c(-1.5,-1.25,-1,0,1,1.25,1.5)) +\n",
    "geom_text_repel(box.padding = 0.5, max.overlaps = 20,xlim = c(-Inf, Inf), ylim = c(-Inf, Inf), size = 7) +\n",
    "ylab(\"- Log of Adjusted P Value\") + xlab(\"Enrichment in either Positive or Negative Different Peakset\") +\n",
    "theme(plot.margin = unit(c(2,5,2,15),\"mm\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a03556f5-c73b-4f79-b712-c356a9ad353e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Open a pdf file\n",
    "pdf(\"plots/cd4na_age_volcano.pdf\", width = 14, height = 8) \n",
    "# 2. Create a plot\n",
    "ggplot(JDF, aes(x = Enrichment, y= mlog10Padj, label=delabel)) + geom_point(aes(color = Group)) + theme_minimal() +\n",
    "geom_vline(xintercept=c(-1, 1), col=\"red\") +\n",
    "scale_color_manual(values = c('#d95f02','#1b9e77')) +\n",
    "geom_hline(yintercept=5, col=\"red\") + ggtitle(\"CISBP Motif Enrichment of CD4Na Age\") +\n",
    "coord_cartesian(clip = \"off\") + scale_x_continuous(trans = squish_trans(-1,1,20),\n",
    "                                                    breaks = c(-1.5,-1.25,-1,0,1,1.25,1.5),\n",
    "                                                    labels = c(-1.5,-1.25,-1,0,1,1.25,1.5)) +\n",
    "geom_text_repel(box.padding = 0.5, max.overlaps = 20,xlim = c(-Inf, Inf), ylim = c(-Inf, Inf), size = 7) +\n",
    "ylab(\"- Log of Adjusted P Value\") + xlab(\"Enrichment in either Positive or Negative Different Peakset\") +\n",
    "theme(plot.margin = unit(c(2,5,2,15),\"mm\"))\n",
    "# Close the pdf file\n",
    "dev.off() "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R 4",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
